---
title: "SNV profile similarity exploration"
subtitle: "210408_psen1_fADfAI_snv"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    theme: flatly
    toc: true
    toc_float: true
    toc_depth: 4
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE, 
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Introduction

# Setup 

```{r packages}
suppressPackageStartupMessages({
  ## Common
  library(tidyverse)
  library(magrittr)
  library(here)
  library(AnnotationHub)
  library(purrr)
  library(scales)
  library(kableExtra)
  library(tictoc)
  library(ggrepel)
  library(RColorBrewer)
  library(ggpubr)
  library(pander)
  library(rmarkdown)
  ## Project specific
  library(VariantAnnotation)
  library(plyranges)
})
```

```{r commonOpts}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
```

## EnsDb

```{r ensParams}
ens_species <- "Danio rerio"
ens_release <- "101"
```

```{r ensDb}
ah <- AnnotationHub() %>%
  subset(species == ens_species) %>%
  subset(rdataclass == "EnsDb")
ahId <- ah$ah_id[str_detect(ah$title, ens_release)]
ensDb <- ah[[ahId]]
```

An `EnsDb` object for Ensembl release 101 was setup for extracting gene and exon annotation information.

## Genome metadata

```{r}
if (ens_species == "Mus musculus") {
  primary_chrs <- c(seq(1:19), "X", "Y")
} else if (ens_species == "Danio rerio") {
  primary_chrs <- c(seq(1:25))
} else if (ens_species == "Homo sapiens") {
  primary_chrs <- c(seq(1:22), "X", "Y")
}
```

```{r ensFeatures}
genes <- genes(ensDb)
mcols(genes) <- mcols(genes)[
  c("gene_id", "gene_name", "gene_biotype", "entrezid")
]
exons <- exonsBy(ensDb, by = "gene")
```

```{r chrLengths}
chrLengths <- seqlengths(ensDb) %>%
  .[primary_chrs]
```

## Sample metadata

```{r}
meta <- read_csv(here("files/samples.csv")) %>%
  dplyr::select(-sample) %>%
  dplyr::rename(sample = basename, genotype = Genotype) %>%
  ## We need some sample aliases that follow R naming conventions
  mutate(
    alias = c(
      paste0(rep("fAD", 7), seq(1, 7)),
      paste0(rep("fAI", 8), seq(1, 8)),
      paste0(rep("wt", 9), seq(1, 9))
    )
  )
meta$genotype <- fct_relevel(
  meta$genotype,
  c("EOfAD-like/+", "fAI-like/+", "WT")
)
```

```{r}
genoCols <- meta$genotype %>%
  unique() %>%
  length() %>%
  brewer.pal("Set1") %>%
  setNames(unique(meta$genotype))
```

# SNV profiles

```{r}
vcf_file <- file.path(
  "/hpcfs/users/a1647910",
  "210408_psen1_fADfAI/analysis-variant_discovery",
  "10_variants/6_select/all_samples.vcf.gz"
)
svp <- ScanVcfParam(info = "", geno = c("GT", "GQ", "DP"))
vcf <- suppressWarnings({
  readVcf(vcf_file, param = svp)
  # readVcf(vcf_file)
})
samples <- samples(header(vcf))
```

```{r}
create_profile <- function(sample, vcf, min_GQ = 20, min_DP = 10, chromosome = NULL){
  gr <- rowRanges(vcf)[,c("REF", "ALT")]
  gr$ALT <- CharacterList(gr$ALT) %>%
    unstrsplit(sep = ",")
  gr$REF <- as.character(gr$REF)
  gr$ALL <- paste(gr$REF, gr$ALT, sep = ",")
  gr$sample <- sample
  gr$GT <- geno(vcf)$GT[,sample] %>%
    str_replace(., "\\|", "\\/")  # We are not concerned with phasing
  gr$GQ <- geno(vcf)$GQ[,sample]
  gr$DP <- geno(vcf)$DP[,sample]
  if (!is.null(chromosome)) {
    gr <- plyranges::filter(gr, seqnames == chromosome)
  }
  gr <- gr %>%
    plyranges::filter(GQ >= min_GQ & DP >= min_DP) %>%
    as.data.frame() %>%
    separate(
      col = "GT", into = c("allele_1", "allele_2"), sep = "/",
      remove = FALSE, convert = TRUE, fill = "right"
    ) %>%
    GRanges()
  gr$allele_1 <- map2(gr$ALL, gr$allele_1, function(x, y){
    unlist(str_split(x, ","))[y + 1]  # GT is 0 based so need to add 1
  }) %>%
    unlist()
  gr$allele_2 <- map2(gr$ALL, gr$allele_2, function(x, y){
    unlist(str_split(x, ","))[y + 1]  # GT is 0 based so need to add 1
  }) %>%
    unlist()
  gr$genotype <- paste(gr$allele_1, gr$allele_2, sep = ":")
  gr
}
```

```{r}
compareProfiles <- function(prof_1, prof_2, chromosome = NULL){
  if (!is.null(chromosome)) {
    prof_1 <- plyranges::filter(prof_1, seqnames == chromosome)
    prof_2 <- plyranges::filter(prof_2, seqnames == chromosome)
  }
  overlaps <- findOverlaps(prof_1, prof_2)
  prof_1 <- prof_1[queryHits(overlaps)]
  prof_2 <- prof_2[subjectHits(overlaps)]
  shared <- GRanges(
    seqnames = seqnames(prof_1),
    ranges = ranges(prof_1),
    strand = strand(prof_1)
  )
  mcols(shared)$sample_1 <- prof_1$sample
  mcols(shared)$sample_2 <- prof_2$sample
  mcols(shared)$allele_1.sample_1 <- mcols(prof_1)$allele_1
  mcols(shared)$allele_1.sample_2 <- mcols(prof_2)$allele_1
  mcols(shared)$allele_2.sample_1 <- mcols(prof_1)$allele_2
  mcols(shared)$allele_2.sample_2 <- mcols(prof_2)$allele_2
  mcols(shared)$genotype.sample_1 <- mcols(prof_1)$genotype
  mcols(shared)$genotype.sample_2 <- mcols(prof_2)$genotype
  mcols(shared)$matches_1 <- 
    mcols(shared)$allele_1.sample_1 == mcols(shared)$allele_1.sample_2
  mcols(shared)$matches_2 <- 
    mcols(shared)$allele_2.sample_1 == mcols(shared)$allele_2.sample_2
  mcols(shared)$matches_sum <-
    mcols(shared)$matches_1 + mcols(shared)$matches_2
  return(shared)
}
```

```{r}
alleleSimilarity <- function(comparison, chromosome = NULL){
  if (!is.null(chromosome)) {
    comparison <- plyranges::filter(comparison, seqnames == chromosome)
  }
  tibble(
    sample_1 = unique(mcols(comparison)$sample_1),
    sample_2 = unique(mcols(comparison)$sample_2),
    allele_matches = sum(mcols(comparison)$matches_sum),
    alleles_total = length(comparison) * 2
  ) %>%
    mutate(
      # similarity = (allele_matches + 1) / (alleles_total + 5)
      similarity = allele_matches / alleles_total
    )
}
```

```{r}
prof_1 <- create_profile(samples[[1]], vcf)
prof_2 <- create_profile(samples[[2]], vcf)
comparison <- compareProfiles(prof_1, prof_2)
similarity <- alleleSimilarity(comparison)
```

```{r}
allPairwiseComparisons <- function(
  samples, vcf, min_GQ = 20, min_DP = 10, chromosome = NULL
) {
  n_total <- length(samples) ^ 2
  n_current <- 1
  lapply(samples, function(sample_1){
    prof_1 <- create_profile(sample_1, vcf, chromosome = chromosome)
    lapply(samples, function(sample_2){
      prof_2 <- create_profile(sample_2, vcf, chromosome = chromosome)
      comparison <- compareProfiles(prof_1, prof_2)
      similarity <- alleleSimilarity(comparison)
      cat(paste0(
        "Finished comparing samples ",
        sample_1,
        " and ",
        sample_2,
        ". [",
        n_current,
        "/",
        n_total,
        "]\n"
      ))
      n_current <<- n_current + 1
      similarity
    }) %>%
      bind_rows()
  }) %>%
    bind_rows()
}
```

```{r}
similarities <- allPairwiseComparisons(samples, vcf)
```

<!--
```{r}
similarities_chr17 <- allPairwiseComparisons(samples, vcf, chromosome = "17")
```

```{r}
similarities_chr16 <- allPairwiseComparisons(samples, vcf, chromosome = "16")
```

```{r}
similarities_chr5 <- allPairwiseComparisons(samples, vcf, chromosome = "5")
```

```{r}
similarities_chr12 <- allPairwiseComparisons(samples, vcf, chromosome = "12")
```

```{r}
sapply(primary_chrs, function(chromosome){
  allPairwiseComparisons(samples, vcf, chromosome = chromosome)
})
```
-->

---

```{r}
similarities_mat <- similarities %>%
  dplyr::select(-allele_matches, -alleles_total) %>%
  pivot_wider(names_from = "sample_2", values_from = "similarity") %>%
  as.data.frame() %>%
  column_to_rownames("sample_1") %>%
  as.matrix() %>%
  .[meta$sample, meta$sample]
similarities_mat[upper.tri(similarities_mat)] <- NA
# similarities_mat[lower.tri(similarities_mat)] <- NA
```

```{r}
library(ComplexHeatmap)
library(circlize)
Heatmap(
  similarities_mat,
  cluster_rows = TRUE,
  cluster_columns = TRUE,
  na_col = "white",
  left_annotation = rowAnnotation(
    genotype = meta$genotype,
    col = list(genotype = genoCols)
  ),
  top_annotation = columnAnnotation(
    genotype = meta$genotype,
    col = list(genotype = genoCols)
  ),
  # row_split = meta$genotype,
  # column_split = meta$genotype,
  column_title = NULL,
  row_title = NULL,
  show_row_names = TRUE,
  show_column_names = TRUE,
  name = "Similarity",  # Names colour legend
  col = colorRamp2(
    breaks = c(0, 0.20, 0.40, 1),
    colors = c("white", "white", "grey", "navy")
  )
)
```