<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lachlan Baer" />


<title>Allele Specific Expression Analysis</title>

<script src="site_libs/header-attrs-2.9/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">210408_psen1_fADfAI_snv</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="de_analysis.html">DE analysis</a>
</li>
<li>
  <a href="de_enrichment.html">DE enrichment</a>
</li>
<li>
  <a href="ase_qc.html">ASE QC</a>
</li>
<li>
  <a href="ase_analysis.html">ASE analysis</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/baerlachlan/210408_psen1_fADfAI_snv">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Allele Specific Expression Analysis</h1>
<h4 class="author">Lachlan Baer</h4>
<h4 class="date">23 September, 2021</h4>

</div>


<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>The motivation for performing an allele specific expression (ASE) analysis stems from the idea that selecting for a particular chromosome (i.e. the mutant chromosome) within an experimental group also selects for any expression quantitative trait loci (eQTLs) that exist on that chromosome. Therefore exists the possibility of differences in transcript abundances between groups tested for differential expression due to different sets of allelic expression characteristics and separate from the effect of the mutation. This would not be a problem in isogenic organisms such as mice, but zebrafish are anything but isogenic. If the eQTL idea is true, one would expect to see a number of differentially expressed genes on the same chromosome of the mutation, which is indeed the case in this dataset. However, if not true, and the observations are caused by the effect of the mutation, an explanation must describe the functional relationship between genes on the same chromosome as the mutation. One such possibility is that functionally related genes tend to cluster on chromosomes through evolutionary time because it allows for advantageous allelic combinations to remain together due to the suppressed chance for separation through recombination or chromosomal segregation.</p>
<p>The aim of this analysis is to therefore tease apart these two possibilities to the best ability with the available RNA-seq data. In particular, allele specific expression testing will be used to assess:</p>
<ol style="list-style-type: decimal">
<li>The potential for ASE to impact differential expression testing in an undesired manner.</li>
<li>Whether an overall enrichment of ASE is observed on the chromosome harbouring the mutation(s).</li>
</ol>
</div>
<div id="setup" class="section level1">
<h1>Setup</h1>
<pre class="r"><code>suppressPackageStartupMessages({
  ## Common
  library(tidyverse)
  library(magrittr)
  library(future.apply)
  library(here)
  library(AnnotationHub)
  library(purrr)
  library(scales)
  library(kableExtra)
  library(tictoc)
  library(ggrepel)
  library(RColorBrewer)
  library(ggpubr)
  library(pander)
  library(rmarkdown)
  ## Project specific
  library(statmod)
  library(fgsea)
})</code></pre>
<pre class="r"><code>if (interactive()) setwd(here::here())
theme_set(theme_bw())
cores &lt;- availableCores() - 2
source(&quot;~/bioinformatics/bioToolkit/lbFuncs.R&quot;)
## Select base directory for the project
projDir &lt;- here()
# projDir &lt;- &quot;/hpcfs/users/a1647910/210408_psen1_fADfAI_snv&quot;</code></pre>
<div id="ensdb" class="section level2">
<h2>EnsDb</h2>
<p>An <code>EnsDb</code> object was obtained for Ensembl release 101 from which gene and exon information was extracted for annotation required in the analysis.</p>
<pre class="r"><code>ah &lt;- AnnotationHub() %&gt;%
  subset(species == &quot;Danio rerio&quot;) %&gt;%
  subset(rdataclass == &quot;EnsDb&quot;)
ensDb &lt;- ah[[&quot;AH83189&quot;]] ## Ens101
genes &lt;- genes(ensDb)
mcols(genes) &lt;- mcols(genes)[
  c(&quot;gene_id&quot;, &quot;gene_name&quot;, &quot;gene_biotype&quot;, &quot;entrezid&quot;)
]
exons &lt;- exonsBy(ensDb, by = &quot;gene&quot;)
drChrs &lt;- seq(1:25)</code></pre>
<p>As the chromosomal localisation of genes is of interest, list and dataframe objects containing gene and chromosome information were generated.</p>
<pre class="r"><code>chrGenes &lt;- genes %&gt;%
  as_tibble() %&gt;%
  dplyr::filter(seqnames %in% drChrs) %&gt;%
  droplevels() %&gt;%
  split(.$seqnames) %&gt;%
  lapply(pull, gene_id) %&gt;%
  set_names(paste0(&quot;chr&quot;, names(.)))</code></pre>
<pre class="r"><code>chrGenes_df &lt;- chrGenes %&gt;%
  lapply(list) %&gt;%
  as_tibble() %&gt;%
  pivot_longer(cols = everything(), names_to = &quot;chr&quot;, values_to= &quot;feat&quot;) %&gt;%
  unnest(feat)</code></pre>
</div>
<div id="metadata" class="section level2">
<h2>Metadata</h2>
<pre class="r"><code>metadata &lt;- read_csv(file.path(projDir, &quot;files/samples.csv&quot;)) %&gt;%
  dplyr::select(-sample) %&gt;%
  dplyr::rename(sample = basename, genotype = Genotype) %&gt;%
  ## We need some sample aliases that follow R naming conventions
  mutate(
    alias = c(
      paste0(rep(&quot;fAD&quot;, 7), seq(1, 7)),
      paste0(rep(&quot;fAI&quot;, 8), seq(1, 8)),
      paste0(rep(&quot;wt&quot;, 9), seq(1, 9))
    )
  )
metadata$genotype &lt;- fct_relevel(
  metadata$genotype,
  c(&quot;EOfAD-like/+&quot;, &quot;fAI-like/+&quot;, &quot;WT&quot;)
)
genoCols &lt;- metadata$genotype %&gt;%
  levels() %&gt;%
  length() %&gt;%
  brewer.pal(&quot;Set1&quot;) %&gt;%
  setNames(levels(metadata$genotype))
compCols &lt;- genoCols[1:2]</code></pre>
<pre class="r"><code>metadata %&gt;%
  dplyr::rename(
    Sample = sample, `Fish ID` = fish_id, `Batch Killed` = batch_killed,
    `Genotype 1` = genotype, `Genotype 2` = Genotype_2, Alias = alias
  ) %&gt;%
  kable(
    align = &quot;l&quot;,
    caption = &quot;Sample metadata&quot;
  ) %&gt;%
  kable_styling(
    bootstrap_options = c(&quot;striped&quot;, &quot;hover&quot;, &quot;condensed&quot;, &quot;responsive&quot;)
  )</code></pre>
<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left: auto; margin-right: auto;">
<caption>
Sample metadata
</caption>
<thead>
<tr>
<th style="text-align:left;">
Sample
</th>
<th style="text-align:left;">
Fish ID
</th>
<th style="text-align:left;">
Batch Killed
</th>
<th style="text-align:left;">
Sex
</th>
<th style="text-align:left;">
RINe
</th>
<th style="text-align:left;">
Genotype 1
</th>
<th style="text-align:left;">
Genotype 2
</th>
<th style="text-align:left;">
Alias
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
7_KB_B12
</td>
<td style="text-align:left;">
27
</td>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD1
</td>
</tr>
<tr>
<td style="text-align:left;">
22_KB_C6
</td>
<td style="text-align:left;">
30
</td>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.6
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD2
</td>
</tr>
<tr>
<td style="text-align:left;">
16_KB_C12
</td>
<td style="text-align:left;">
40
</td>
<td style="text-align:left;">
8
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD3
</td>
</tr>
<tr>
<td style="text-align:left;">
8_KB_B3
</td>
<td style="text-align:left;">
46
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.2
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD4
</td>
</tr>
<tr>
<td style="text-align:left;">
19_KB_C3
</td>
<td style="text-align:left;">
29
</td>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.5
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD5
</td>
</tr>
<tr>
<td style="text-align:left;">
10_KB_B6
</td>
<td style="text-align:left;">
50
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD6
</td>
</tr>
<tr>
<td style="text-align:left;">
4_KB_A6
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
3
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.0
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD7
</td>
</tr>
<tr>
<td style="text-align:left;">
2_KB_A11
</td>
<td style="text-align:left;">
12
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
8.7
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI1
</td>
</tr>
<tr>
<td style="text-align:left;">
12_KB_B8
</td>
<td style="text-align:left;">
22
</td>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI2
</td>
</tr>
<tr>
<td style="text-align:left;">
15_KB_C11
</td>
<td style="text-align:left;">
45
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI3
</td>
</tr>
<tr>
<td style="text-align:left;">
18_KB_C2
</td>
<td style="text-align:left;">
47
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.2
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI4
</td>
</tr>
<tr>
<td style="text-align:left;">
9_KB_B5
</td>
<td style="text-align:left;">
21
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI5
</td>
</tr>
<tr>
<td style="text-align:left;">
6_KB_B11
</td>
<td style="text-align:left;">
26
</td>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI6
</td>
</tr>
<tr>
<td style="text-align:left;">
24_KB_C8
</td>
<td style="text-align:left;">
44
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.8
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI7
</td>
</tr>
<tr>
<td style="text-align:left;">
21_KB_C5
</td>
<td style="text-align:left;">
49
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.5
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI8
</td>
</tr>
<tr>
<td style="text-align:left;">
5_KB_A7
</td>
<td style="text-align:left;">
19
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.1
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt1
</td>
</tr>
<tr>
<td style="text-align:left;">
1_KB_A10
</td>
<td style="text-align:left;">
25
</td>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt2
</td>
</tr>
<tr>
<td style="text-align:left;">
23_KB_C7
</td>
<td style="text-align:left;">
36
</td>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.7
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt3
</td>
</tr>
<tr>
<td style="text-align:left;">
3_KB_A1
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt4
</td>
</tr>
<tr>
<td style="text-align:left;">
17_KB_C1
</td>
<td style="text-align:left;">
34
</td>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.0
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt5
</td>
</tr>
<tr>
<td style="text-align:left;">
20_KB_C4
</td>
<td style="text-align:left;">
35
</td>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt6
</td>
</tr>
<tr>
<td style="text-align:left;">
14_KB_C10
</td>
<td style="text-align:left;">
37
</td>
<td style="text-align:left;">
8
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt7
</td>
</tr>
<tr>
<td style="text-align:left;">
13_KB_B9
</td>
<td style="text-align:left;">
54
</td>
<td style="text-align:left;">
10
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt8
</td>
</tr>
<tr>
<td style="text-align:left;">
11_KB_B7
</td>
<td style="text-align:left;">
55
</td>
<td style="text-align:left;">
10
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.1
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt9
</td>
</tr>
</tbody>
</table>
<pre class="r"><code>genotypes &lt;- unique(metadata$genotype) %&gt;%
  as.vector() %&gt;%
  set_names(., .)
fad &lt;- metadata$alias %&gt;%
  str_detect(&quot;fAD&quot;)
fai &lt;- metadata$alias %&gt;%
  str_detect(&quot;fAI&quot;)
wt &lt;- metadata$alias %&gt;%
  str_detect(&quot;WT&quot;)</code></pre>
</div>
</div>
<div id="allele-specific-expression" class="section level1">
<h1>Allele specific expression</h1>
<p>Allele specific expression testing was performed for each sample at the gene-level with <code>GeneiASE</code> software. Genes were classified as showing allele specific expression (ASE) if they had a FDR-adjusted p-value &lt; 0.05.</p>
<pre class="r"><code>files &lt;- list.files(
  file.path(projDir, &quot;13_geneiase/2_ase&quot;),
  full.names = TRUE
)
samples &lt;- basename(files) %&gt;%
  str_remove(&quot;.static.pval.tsv&quot;)
topASE &lt;- lapply(files, function(x){
  sample &lt;- basename(x) %&gt;%
    str_remove(&quot;.static.pval.tsv&quot;)
  read_tsv(x) %&gt;%
    mutate(
      sig = fdr &lt; 0.05,
      sample = sample
    ) %&gt;%
    left_join(metadata[,c(&quot;sample&quot;, &quot;alias&quot;, &quot;genotype&quot;)]) %&gt;%
    dplyr::arrange(fdr) %&gt;%
    mutate(ASE = fdr &lt; 0.05)
}) %&gt;%
  set_names(metadata$alias[match(samples, metadata$sample)]) %&gt;%
  .[metadata$alias]</code></pre>
<pre class="r"><code>aseRC &lt;- readRDS(here(&quot;files/aseRC.Rds&quot;))
aseRCByGene &lt;- readRDS(here(&quot;files/aseRCByGene.Rds&quot;)) %&gt;%
  sapply(dplyr::select, gene, everything(), simplify = FALSE)</code></pre>
<pre class="r"><code>nASE &lt;- bind_rows(topASE) %&gt;%
  group_by(alias, sample, genotype) %&gt;%
  summarise(nASE = sum(ASE), total = n()) %&gt;%
  ungroup() %&gt;%
  mutate(
    alias = factor(alias),
    propASE = nASE / total
  )</code></pre>
<p>Between <em>327</em> and <em>557</em> genes were classified as showing ASE across all samples, which accounted for between <em>4.53%</em> and <em>7.14%</em> of the total genes tested.</p>
<pre class="r"><code>nASE %&gt;%
  ggplot(aes(alias, nASE, fill = genotype)) +
  geom_bar(stat = &quot;identity&quot;, colour = &quot;black&quot;) +
  scale_fill_manual(values = genoCols) +
  scale_y_continuous(breaks = seq(0, 500, 100)) +
  labs(
    x = &quot;Sample&quot;,
    y = &quot;Number of ASE genes&quot;,
    fill = &quot;Genotype&quot;
  ) +
  theme(
    axis.text.x = element_text(angle = 270, vjust = 0.5, hjust = 0)
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ase_analysis_files/figure-html/nAsePlot-1.png" alt="*The number of genes classified as showing significant ASE basedon an FDR-adjusted p-value cut-off of 0.05. The number of genes showing ASE was fairly consistent across all samples and genotypes.*" width="768" />
<p class="caption">
<em>The number of genes classified as showing significant ASE basedon an FDR-adjusted p-value cut-off of 0.05. The number of genes showing ASE was fairly consistent across all samples and genotypes.</em>
</p>
</div>
</div>
<div id="ase-overlap-with-de-genes" class="section level1">
<h1>ASE overlap with DE genes</h1>
<p>Previously analysed different expression testing results were loaded as a list of topTables.</p>
<pre class="r"><code>topDE &lt;- readRDS(file.path(projDir, &quot;files/topTables.Rds&quot;)) %&gt;%
  set_names(c(&quot;fAD&quot;, &quot;fAI&quot;))</code></pre>
<p>The fAD vs WT comparison showed 10 DE genes.</p>
<pre class="r"><code>topDE[[1]] %&gt;%
  dplyr::filter(DE) %&gt;%
  dplyr::select(geneID = gene_id, symbol = gene_name, chromosome, logFC, logCPM, FDR, bonfP) %&gt;%
  mutate(
    logCPM = formatC(logCPM, digits = 2, format = &quot;f&quot;),
    logFC = formatC(logFC, digits = 2, format = &quot;f&quot;),
    FDR = formatC(FDR, digits = 2, format = &quot;e&quot;),
    bonfP = formatC(bonfP, digits = 2, format = &quot;e&quot;),
  ) %&gt;%
  kable(
    align = &quot;l&quot;,
    caption = paste(
      &quot;The&quot;, nrow(.), &quot;differentially expressed genes for Alzheimer&#39;s genotype&quot;,
      &quot;in comparison to wildtype&quot;
    )
  ) %&gt;%
  kable_styling(
    bootstrap_options = c(&quot;striped&quot;, &quot;hover&quot;, &quot;condensed&quot;, &quot;responsive&quot;)
  )</code></pre>
<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left: auto; margin-right: auto;">
<caption>
The 10 differentially expressed genes for Alzheimer’s genotype in comparison to wildtype
</caption>
<thead>
<tr>
<th style="text-align:left;">
geneID
</th>
<th style="text-align:left;">
symbol
</th>
<th style="text-align:left;">
chromosome
</th>
<th style="text-align:left;">
logFC
</th>
<th style="text-align:left;">
logCPM
</th>
<th style="text-align:left;">
FDR
</th>
<th style="text-align:left;">
bonfP
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
ENSDARG00000078322
</td>
<td style="text-align:left;">
col12a1a
</td>
<td style="text-align:left;">
17
</td>
<td style="text-align:left;">
1.87
</td>
<td style="text-align:left;">
3.83
</td>
<td style="text-align:left;">
1.73e-16
</td>
<td style="text-align:left;">
1.73e-16
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000093378
</td>
<td style="text-align:left;">
si:ch211-235i11.5
</td>
<td style="text-align:left;">
17
</td>
<td style="text-align:left;">
0.63
</td>
<td style="text-align:left;">
2.96
</td>
<td style="text-align:left;">
2.56e-07
</td>
<td style="text-align:left;">
5.11e-07
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000079324
</td>
<td style="text-align:left;">
ano9b
</td>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
1.37
</td>
<td style="text-align:left;">
-1.19
</td>
<td style="text-align:left;">
2.35e-04
</td>
<td style="text-align:left;">
7.05e-04
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000020448
</td>
<td style="text-align:left;">
adi1
</td>
<td style="text-align:left;">
17
</td>
<td style="text-align:left;">
0.54
</td>
<td style="text-align:left;">
2.81
</td>
<td style="text-align:left;">
1.69e-03
</td>
<td style="text-align:left;">
6.74e-03
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000092191
</td>
<td style="text-align:left;">
CR318588.1
</td>
<td style="text-align:left;">
18
</td>
<td style="text-align:left;">
3.31
</td>
<td style="text-align:left;">
2.45
</td>
<td style="text-align:left;">
5.15e-03
</td>
<td style="text-align:left;">
2.68e-02
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000001993
</td>
<td style="text-align:left;">
myhb
</td>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
4.04
</td>
<td style="text-align:left;">
2.38
</td>
<td style="text-align:left;">
5.15e-03
</td>
<td style="text-align:left;">
3.09e-02
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000088298
</td>
<td style="text-align:left;">
si:ch211-235i11.4
</td>
<td style="text-align:left;">
17
</td>
<td style="text-align:left;">
0.34
</td>
<td style="text-align:left;">
3.57
</td>
<td style="text-align:left;">
7.51e-03
</td>
<td style="text-align:left;">
5.26e-02
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000094719
</td>
<td style="text-align:left;">
CR318588.3
</td>
<td style="text-align:left;">
18
</td>
<td style="text-align:left;">
2.05
</td>
<td style="text-align:left;">
2.71
</td>
<td style="text-align:left;">
1.18e-02
</td>
<td style="text-align:left;">
9.46e-02
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000088507
</td>
<td style="text-align:left;">
znf982
</td>
<td style="text-align:left;">
17
</td>
<td style="text-align:left;">
0.39
</td>
<td style="text-align:left;">
2.28
</td>
<td style="text-align:left;">
4.94e-02
</td>
<td style="text-align:left;">
4.91e-01
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000056065
</td>
<td style="text-align:left;">
mov10b.2
</td>
<td style="text-align:left;">
8
</td>
<td style="text-align:left;">
1.37
</td>
<td style="text-align:left;">
-1.78
</td>
<td style="text-align:left;">
4.94e-02
</td>
<td style="text-align:left;">
4.94e-01
</td>
</tr>
</tbody>
</table>
<p>The fAI vs WT comparison showed 6 DE genes.</p>
<pre class="r"><code>topDE[[2]] %&gt;%
  dplyr::filter(DE) %&gt;%
  dplyr::select(geneID = gene_id, symbol = gene_name, chromosome, logFC, logCPM, FDR, bonfP) %&gt;%
  mutate(
    logCPM = formatC(logCPM, digits = 2, format = &quot;f&quot;),
    logFC = formatC(logFC, digits = 2, format = &quot;f&quot;),
    FDR = formatC(FDR, digits = 2, format = &quot;e&quot;),
    bonfP = formatC(bonfP, digits = 2, format = &quot;e&quot;),
  ) %&gt;%
  kable(
    align = &quot;l&quot;,
    caption = paste(
      &quot;The&quot;, nrow(.), &quot;differentially expressed genes for Acne Inversa genotype&quot;,
      &quot;in comparison to wildtype&quot;
    )
  ) %&gt;%
  kable_styling(
    bootstrap_options = c(&quot;striped&quot;, &quot;hover&quot;, &quot;condensed&quot;, &quot;responsive&quot;)
  )</code></pre>
<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left: auto; margin-right: auto;">
<caption>
The 6 differentially expressed genes for Acne Inversa genotype in comparison to wildtype
</caption>
<thead>
<tr>
<th style="text-align:left;">
geneID
</th>
<th style="text-align:left;">
symbol
</th>
<th style="text-align:left;">
chromosome
</th>
<th style="text-align:left;">
logFC
</th>
<th style="text-align:left;">
logCPM
</th>
<th style="text-align:left;">
FDR
</th>
<th style="text-align:left;">
bonfP
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
ENSDARG00000004870
</td>
<td style="text-align:left;">
psen1
</td>
<td style="text-align:left;">
17
</td>
<td style="text-align:left;">
-0.78
</td>
<td style="text-align:left;">
5.33
</td>
<td style="text-align:left;">
4.11e-70
</td>
<td style="text-align:left;">
4.11e-70
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000005179
</td>
<td style="text-align:left;">
dglucy
</td>
<td style="text-align:left;">
17
</td>
<td style="text-align:left;">
-0.97
</td>
<td style="text-align:left;">
5.01
</td>
<td style="text-align:left;">
3.84e-05
</td>
<td style="text-align:left;">
7.68e-05
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000061758
</td>
<td style="text-align:left;">
sh3pxd2ab
</td>
<td style="text-align:left;">
17
</td>
<td style="text-align:left;">
-3.04
</td>
<td style="text-align:left;">
1.06
</td>
<td style="text-align:left;">
2.03e-04
</td>
<td style="text-align:left;">
6.08e-04
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000067665
</td>
<td style="text-align:left;">
fam167aa
</td>
<td style="text-align:left;">
17
</td>
<td style="text-align:left;">
-3.40
</td>
<td style="text-align:left;">
0.32
</td>
<td style="text-align:left;">
3.51e-04
</td>
<td style="text-align:left;">
1.40e-03
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000105453
</td>
<td style="text-align:left;">
kcnk17
</td>
<td style="text-align:left;">
17
</td>
<td style="text-align:left;">
-1.52
</td>
<td style="text-align:left;">
-1.00
</td>
<td style="text-align:left;">
3.23e-03
</td>
<td style="text-align:left;">
1.61e-02
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000094098
</td>
<td style="text-align:left;">
si:ch211-278p9.3
</td>
<td style="text-align:left;">
17
</td>
<td style="text-align:left;">
-2.29
</td>
<td style="text-align:left;">
-2.34
</td>
<td style="text-align:left;">
4.28e-02
</td>
<td style="text-align:left;">
2.57e-01
</td>
</tr>
</tbody>
</table>
<p>For each differentially expressed gene, ASE results were assessed across each sample. Due to the requirements for each gene to possess at least one heterozygous single nucleotide variant (SNV) in order to be tested for ASE, not all detected genes were subject to the analysis. Furthermore, a substantial number of genes containing het-SNVs were excluded from ASE testing as they did not pass quality control requirements. Therefore it is important to distinguish whether a gene was tested and did not show ASE opposed to not being tested at all. For each DE gene, samples were assigned to one of three categories:</p>
<ol style="list-style-type: decimal">
<li>Tested and shows ASE</li>
<li>Tested and does not show ASE</li>
<li>Not tested (no detected het-SNVs or filtered by QC)</li>
</ol>
<pre class="r"><code>deGenes &lt;- sapply(topDE, dplyr::filter, DE, simplify = FALSE)</code></pre>
<pre class="r"><code>deAseOverlap &lt;- sapply(deGenes, function(x){
  lapply(topASE, function(y){
    dplyr::select(x, feat = gene_id, symbol = gene_name, chromosome) %&gt;%
      left_join(y[,c(&quot;feat&quot;, &quot;sample&quot;, &quot;genotype&quot;, &quot;ASE&quot;)])
  }) %&gt;%
    bind_rows() %&gt;%
    group_by(feat, ASE) %&gt;%
    tally() %&gt;%
    ungroup() %&gt;%
    pivot_wider(names_from = ASE, values_from = n, values_fill = 0) %&gt;%
    dplyr::select(gene = feat, ASE = `TRUE`, no_ASE = `FALSE`, not_tested = `NA`)
}, simplify = FALSE)</code></pre>
<p><strong>Alzheimer’s DE genes:</strong></p>
<pre class="r"><code>deAseOverlap$fAD %&gt;%
  kable(
    align = &quot;l&quot;,
    caption = paste(
      &quot;The number of samples that were tested and showed ASE, tested and did&quot;,
      &quot;not show ASE, or not tested at all for each differentially expressed&quot;,
      &quot;gene in the Alzheimer&#39;s vs wildtype comparison&quot;
    )
  ) %&gt;%
  kable_styling(
    bootstrap_options = c(&quot;striped&quot;, &quot;hover&quot;, &quot;condensed&quot;, &quot;responsive&quot;)
  )</code></pre>
<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left: auto; margin-right: auto;">
<caption>
The number of samples that were tested and showed ASE, tested and did not show ASE, or not tested at all for each differentially expressed gene in the Alzheimer’s vs wildtype comparison
</caption>
<thead>
<tr>
<th style="text-align:left;">
gene
</th>
<th style="text-align:left;">
ASE
</th>
<th style="text-align:left;">
no_ASE
</th>
<th style="text-align:left;">
not_tested
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
ENSDARG00000001993
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
24
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000020448
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
16
</td>
<td style="text-align:left;">
8
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000056065
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
24
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000078322
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
18
</td>
<td style="text-align:left;">
6
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000079324
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
24
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000088298
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
8
</td>
<td style="text-align:left;">
16
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000088507
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
8
</td>
<td style="text-align:left;">
16
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000092191
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
11
</td>
<td style="text-align:left;">
12
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000093378
</td>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
3
</td>
<td style="text-align:left;">
16
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000094719
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
23
</td>
<td style="text-align:left;">
1
</td>
</tr>
</tbody>
</table>
<p><strong>Acne Inversa DE genes:</strong></p>
<pre class="r"><code>deAseOverlap$fAI %&gt;%
  kable(
    align = &quot;l&quot;,
    caption = paste(
      &quot;The number of samples that were tested and showed ASE, tested and did&quot;,
      &quot;not show ASE, or not tested at all for each differentially expressed&quot;,
      &quot;gene in the Acne Inversa vs wildtype comparison&quot;
    )
  ) %&gt;%
  kable_styling(
    bootstrap_options = c(&quot;striped&quot;, &quot;hover&quot;, &quot;condensed&quot;, &quot;responsive&quot;)
  )</code></pre>
<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left: auto; margin-right: auto;">
<caption>
The number of samples that were tested and showed ASE, tested and did not show ASE, or not tested at all for each differentially expressed gene in the Acne Inversa vs wildtype comparison
</caption>
<thead>
<tr>
<th style="text-align:left;">
gene
</th>
<th style="text-align:left;">
ASE
</th>
<th style="text-align:left;">
no_ASE
</th>
<th style="text-align:left;">
not_tested
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
ENSDARG00000004870
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
16
</td>
<td style="text-align:left;">
8
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000005179
</td>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
15
</td>
<td style="text-align:left;">
4
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000061758
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
24
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000067665
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
24
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000094098
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
24
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSDARG00000105453
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
24
</td>
</tr>
</tbody>
</table>
<p>Unfortunately due to the large amounts of samples that could not be examined for overlap of DE and ASE data, no conclusions could be made regarding the ability for ASE to affect DE results.</p>
</div>
<div id="enrichment-testing" class="section level1">
<h1>Enrichment testing</h1>
<p>While no conclusions were achieved to address the potential for ASE to impact differential expression testing in an undesired manner, more general patterns of ASE were assessed through enrichment testing. Two different methods were set up to analyse if ASE is enriched on the chromosome that was selected for:</p>
<ul>
<li>Fisher’s exact test</li>
<li>Gene set enrichment analysis (GSEA)</li>
</ul>
<div id="fishers-exact-test" class="section level2">
<h2>Fisher’s exact test</h2>
<p>For each chromosome within each sample, the entire set of detected genes tested for ASE were classified by belonging to the chromosome and showing ASE.</p>
<pre class="r"><code>fisherRes &lt;- bind_rows(topASE) %&gt;%
  left_join(chrGenes_df) %&gt;%
  group_by(sample, chr) %&gt;%
  summarise(ase_chr = sum(ASE), noAse_chr = sum(!ASE)) %&gt;%
  group_by(sample) %&gt;%
  mutate(
    ase_notChr = sum(ase_chr) - ase_chr,
    noAse_notChr = sum(noAse_chr) - noAse_chr
  ) %&gt;%
  ungroup() %&gt;%
  mutate(
    fisherTest = pmap(
      ## Setup list input to pmap in order that will construct a correct matrix
      list(ase_chr, ase_notChr, noAse_chr, noAse_notChr),
      ~ fisher.test(matrix(c(..1, ..2, ..3, ..4), ncol = 2))
    ),
    fisher.p = vapply(
      fisherTest,
      function(x){x$p.value},
      numeric(1)
    ),
    fdr.p = p.adjust(fisher.p, method = &quot;fdr&quot;),
    bonf.p = p.adjust(fisher.p, method = &quot;bonf&quot;)
  ) %&gt;%
  left_join(metadata[,c(&quot;sample&quot;, &quot;genotype&quot;)])</code></pre>
<p>This was represented as a 2x2 contingency table for enrichment testing using Fisher’s exact test. An example of a contingency table for chromosome 17 of sample 7_KB_B12 (EOfAD-like/+ genotype) is shown below:</p>
<center>
<pre class="r"><code>dplyr::filter(fisherRes, sample == &quot;7_KB_B12&quot; &amp; chr == &quot;chr17&quot;) %&gt;%
  with(
    matrix(c(ase_chr, ase_notChr, noAse_chr, noAse_notChr), ncol = 2)
  ) %&gt;%
  set_rownames(c(&quot;Chr 17&quot;, &quot;Not Chr 17&quot;)) %&gt;%
  set_colnames(c(&quot;ASE&quot;, &quot;No ASE&quot;)) %&gt;%
  pander()</code></pre>
<table style="width:44%;">
<colgroup>
<col width="23%" />
<col width="8%" />
<col width="12%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">ASE</th>
<th align="center">No ASE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>Chr 17</strong></td>
<td align="center">17</td>
<td align="center">323</td>
</tr>
<tr class="even">
<td align="center"><strong>Not Chr 17</strong></td>
<td align="center">453</td>
<td align="center">7181</td>
</tr>
</tbody>
</table>
</center>
<pre class="r"><code>fisherRes %&gt;%
  mutate(
    sample = factor(sample, levels = metadata$sample),
    chr = factor(chr, levels = str_sort(unique(chr), numeric = TRUE))
  ) %&gt;%
  ggplot(aes(sample, -log10(fdr.p), fill = genotype)) +
  geom_bar(stat = &quot;identity&quot;, position = &quot;dodge&quot;, colour= &quot;black&quot;, width = 0.7) +
  geom_hline(yintercept = -log10(0.05), linetype = &quot;dashed&quot;, ) +
  scale_fill_manual(values = genoCols) +
  labs(
    title = &quot;Chromosomal enrichment of genes showing ASE&quot;,
    x = &quot;Sample&quot;,
    y = log10plab,
    fill = &quot;Genotype&quot;) +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    legend.position = &quot;bottom&quot;
    ) +
  facet_wrap(~chr)</code></pre>
<div class="figure" style="text-align: center">
<img src="ase_analysis_files/figure-html/fisherTestPlot-1.png" alt="*Results from Fisher's exact test for enrichment of ASE genes by chromosome. The -log10 FDR-adjusted p-value is show for each sample, which are coloured by genotype. Each chromosome is faceted for simpler viewing. The dotted line represents a FDR-adjust p-value of 0.05.*" width="768" />
<p class="caption">
<em>Results from Fisher’s exact test for enrichment of ASE genes by chromosome. The -log10 FDR-adjusted p-value is show for each sample, which are coloured by genotype. Each chromosome is faceted for simpler viewing. The dotted line represents a FDR-adjust p-value of 0.05.</em>
</p>
</div>
</div>
<div id="gene-set-enrichment-analysis" class="section level2">
<h2>Gene Set Enrichment Analysis</h2>
<p>As an alternative approach, GSEA was performed using gene sets defined based on the chromosome each gene is located.</p>
<p>The first step for performing GSEA involved creating a ranked list . The methodology that <code>GeneiASE</code> uses for calculating p-values posed two issues when constructing a ranked list:</p>
<ol style="list-style-type: decimal">
<li>Some p-values were 0
<ul>
<li>this causes issues when using -log10(p) as a stat for ranking the genes</li>
</ul></li>
<li>Sets of genes had equal p-values
<ul>
<li>this causes problems the ordering of these genes is arbitrary</li>
</ul></li>
</ol>
<pre class="r"><code>wilsonConf &lt;- function(p.vals, conf.level = 0.95){
  p &lt;- mean(p.vals)
  n &lt;- length(p.vals)
  alpha &lt;- (1 - conf.level)
  alpha2 &lt;- alpha / 2
  z &lt;- qnorm(1 - alpha2)
  z2 &lt;- z * z
  p1 &lt;- p + 0.5 * z2/n
  p2 &lt;- z * sqrt((p * (1 - p) + 0.25 * z2/n)/n)
  p3 &lt;- 1 + z2/n
  lcl &lt;- (p1 - p2)/p3
  ucl &lt;- (p1 + p2)/p3
  res.wilson &lt;- tibble(
    mean.p = p,
    lower = lcl,
    upper = ucl,
    n = n,
    sig = mean.p &lt; alpha
  )
  return(res.wilson)
}</code></pre>
<pre class="r"><code>bonfConf &lt;- function(adj.p.val, n, n.bonf, conf.level = 0.95, limit = c(&quot;lower&quot;, &quot;upper&quot;)){
  alpha &lt;- ((1 - conf.level) / n.bonf)
  alpha2 &lt;- alpha / 2
  z &lt;- qnorm(1 - alpha2)
  z2 &lt;- z * z
  p1 &lt;- adj.p.val + 0.5 * z2/n
  p2 &lt;- z * sqrt((adj.p.val * (1 - adj.p.val) + 0.25 * z2/n)/n)
  p3 &lt;- 1 + z2/n
  lcl &lt;- (p1 - p2)/p3
  ucl &lt;- (p1 + p2)/p3
  if (limit == &quot;lower&quot;){return(lcl)}
  if (limit == &quot;upper&quot;){return(ucl)}
}</code></pre>
<pre class="r"><code>permRanks &lt;- function(topASE){
  topASE[,c(&quot;feat&quot;, &quot;fdr&quot;)] %&gt;%
    mutate(
      fdr = ifelse(fdr == 0, 0.000001, fdr),
      stat = -log10(fdr)
    ) %&gt;%
    split(f = .$stat) %&gt;%
    lapply(function(x){
      pull(x, feat) %&gt;%
        sample(length(.), replace = FALSE)
    }) %&gt;%
    unlist(use.names = FALSE) %&gt;%
    enframe(name = NULL, value = &quot;geneid&quot;) %&gt;%
    mutate(stat = seq(1:nrow(.))) %&gt;%
    with(structure(stat, names = geneid)) %&gt;%
    rev()
}</code></pre>
<pre class="r"><code>perms &lt;- 500
fgseaFile &lt;- paste0(
  here(),
  &quot;/files/fgsea_&quot;,
  perms,
  &quot;perms.Rds&quot;
)</code></pre>
<pre class="r"><code>## This code chunk has a lengthy run-time and will not run if the object has already been saved
## Due to its large size the object remains on the HPC system
## Across 6 threads it takes 1-2 hours to complete
if (!file.exists(fgseaFile)) {
  plan(list(tweak(multisession, workers = cores), sequential))
  ranks &lt;- future_sapply(1:perms, function(...){
    future_sapply(topASE, permRanks, future.seed = 17)
  }, future.seed = 17) %&gt;%
    t() %&gt;%
    as_tibble()
  fgsea &lt;- future_lapply(as.list(ranks), function(x){
    future_lapply(x, function(y){
      fgseaMultilevel(chrGenes, y, scoreType = &quot;pos&quot;, eps = 0)
    }, future.seed = 17)
  }, future.seed = 17) %&gt;%
    set_names(colnames(ranks))
  saveRDS(fgsea, fgseaFile)
  plan(sequential)
}</code></pre>
<pre class="r"><code>fgsea &lt;- readRDS(fgseaFile)</code></pre>
<pre class="r"><code>pvals &lt;- lapply(fgsea, function(x){
  lapply(seq_along(x), function(y){
    x[[y]][,c(&quot;pathway&quot;, &quot;padj&quot;)] %&gt;%
      dplyr::rename(!!as.character(y) := padj)
  }) %&gt;%
    purrr::reduce(left_join) %&gt;%
    column_to_rownames(&quot;pathway&quot;) %&gt;%
    t() %&gt;%
    as_tibble()
})</code></pre>
<pre class="r"><code>mean &lt;- lapply(pvals, function(sample){
  chrList &lt;- as.list(sample)
  lapply(seq_along(chrList), function(x){
    chr &lt;- names(chrList)[x]
    wilsonConf(chrList[[x]]) %&gt;%
      mutate(chromosome = as.factor(chr))
  }) %&gt;%
    bind_rows() %&gt;%
    dplyr::select(chromosome, everything()) %&gt;%
    mutate(
      adj.mean.p = p.adjust(mean.p, method = &quot;bonf&quot;),
      adj.lower = bonfConf(adj.mean.p, n, length(adj.mean.p), limit = &quot;lower&quot;),
      adj.upper = bonfConf(adj.mean.p, n, length(adj.mean.p), limit = &quot;upper&quot;)
    )
}) </code></pre>
<pre class="r"><code>lapply(seq_along(mean), function(ind){
  x &lt;- mean[[ind]]
  x %&gt;%
    mutate(
      sample = names(mean)[ind],
      chromosome = fct_relevel(chromosome, str_sort(chromosome, numeric = TRUE))
    )
}) %&gt;%
  bind_rows() %&gt;%
  left_join(metadata[,c(&quot;alias&quot;, &quot;genotype&quot;)], by = c(&quot;sample&quot; = &quot;alias&quot;)) %&gt;%
  ggplot(aes(sample, -log10(adj.mean.p))) +
  geom_bar(stat = &quot;identity&quot;, aes(fill = genotype)) +
  geom_errorbar(aes(
    ymax = -log10(adj.lower),
    ymin = -log10(adj.upper)
  ), width = 0.2) +
  geom_hline(yintercept = -log10(0.05), linetype = &quot;dashed&quot;) +
  facet_wrap(~ chromosome) +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    legend.position = &quot;bottom&quot;
  ) +
  scale_fill_manual(values = genoCols) +
  labs(y = log10plab)</code></pre>
<p><img src="ase_analysis_files/figure-html/unnamed-chunk-10-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
