<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lachlan Baer" />


<title>Allele Specific Expression Analysis</title>

<script src="site_libs/header-attrs-2.9/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">210408_psen1_fADfAI_snv</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="de_analysis.html">DE analysis</a>
</li>
<li>
  <a href="de_enrichment.html">DE enrichment</a>
</li>
<li>
  <a href="ase_qc.html">ASE QC</a>
</li>
<li>
  <a href="ase_analysis.html">ASE analysis</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/baerlachlan/210408_psen1_fADfAI_snv">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Allele Specific Expression Analysis</h1>
<h4 class="author">Lachlan Baer</h4>
<h4 class="date">22 September, 2021</h4>

</div>


<div id="setup" class="section level1">
<h1>Setup</h1>
<pre class="r"><code>suppressPackageStartupMessages({
  ## Common
  library(tidyverse)
  library(magrittr)
  library(future.apply)
  library(here)
  library(AnnotationHub)
  library(purrr)
  library(scales)
  library(kableExtra)
  library(tictoc)
  library(ggrepel)
  library(RColorBrewer)
  library(ggpubr)
  library(pander)
  library(rmarkdown)
  ## Project specific
  library(statmod)
  library(fgsea)
})</code></pre>
<pre class="r"><code>if (interactive()) setwd(here::here())
theme_set(theme_bw())
cores &lt;- availableCores() - 2</code></pre>
<pre class="r"><code>source(&quot;~/bioinformatics/bioToolkit/lbFuncs.R&quot;)</code></pre>
<pre class="r"><code>## Select base directory for the project
projDir &lt;- here()
# projDir &lt;- &quot;/hpcfs/users/a1647910/210408_psen1_fADfAI_snv&quot;</code></pre>
<div id="ensdb" class="section level2">
<h2>EnsDb</h2>
<pre class="r"><code>ah &lt;- AnnotationHub() %&gt;%
  subset(species == &quot;Danio rerio&quot;) %&gt;%
  subset(rdataclass == &quot;EnsDb&quot;)
ensDb &lt;- ah[[&quot;AH83189&quot;]] ## Ens101
genes &lt;- genes(ensDb)
mcols(genes) &lt;- mcols(genes)[
  c(&quot;gene_id&quot;, &quot;gene_name&quot;, &quot;gene_biotype&quot;, &quot;entrezid&quot;)
]
exons &lt;- exonsBy(ensDb, by = &quot;gene&quot;)
drChrs &lt;- seq(1:25)</code></pre>
<pre class="r"><code>chrGenes &lt;- genes %&gt;%
  as_tibble() %&gt;%
  dplyr::filter(seqnames %in% drChrs) %&gt;%
  droplevels() %&gt;%
  split(.$seqnames) %&gt;%
  lapply(pull, gene_id) %&gt;%
  set_names(paste0(&quot;chr&quot;, names(.)))</code></pre>
<pre class="r"><code>chrGenes_df &lt;- chrGenes %&gt;%
  lapply(list) %&gt;%
  as_tibble() %&gt;%
  pivot_longer(cols = everything(), names_to = &quot;chr&quot;, values_to= &quot;feat&quot;) %&gt;%
  unnest(feat)</code></pre>
</div>
</div>
<div id="load-data" class="section level1">
<h1>Load data</h1>
<div id="metadata" class="section level2">
<h2>Metadata</h2>
<pre class="r"><code>metadata &lt;- read_csv(file.path(projDir, &quot;files/samples.csv&quot;)) %&gt;%
  dplyr::select(-sample) %&gt;%
  dplyr::rename(sample = basename, genotype = Genotype) %&gt;%
  ## We need some sample aliases that follow R naming conventions
  mutate(
    alias = c(
      paste0(rep(&quot;fAD&quot;, 7), seq(1, 7)),
      paste0(rep(&quot;fAI&quot;, 8), seq(1, 8)),
      paste0(rep(&quot;wt&quot;, 9), seq(1, 9))
    )
  )
metadata$genotype &lt;- fct_relevel(
  metadata$genotype,
  c(&quot;EOfAD-like/+&quot;, &quot;fAI-like/+&quot;, &quot;WT&quot;)
)
genoCols &lt;- metadata$genotype %&gt;%
  levels() %&gt;%
  length() %&gt;%
  brewer.pal(&quot;Set1&quot;) %&gt;%
  setNames(levels(metadata$genotype))
compCols &lt;- genoCols[1:2]</code></pre>
<pre class="r"><code>metadata %&gt;%
  dplyr::rename(
    Sample = sample, `Fish ID` = fish_id, `Batch Killed` = batch_killed,
    `Genotype 1` = genotype, `Genotype 2` = Genotype_2, Alias = alias
  ) %&gt;%
  kable(
    align = &quot;l&quot;,
    caption = &quot;Sample metadata&quot;
  ) %&gt;%
  kable_styling(
    bootstrap_options = c(&quot;striped&quot;, &quot;hover&quot;, &quot;condensed&quot;, &quot;responsive&quot;)
  )</code></pre>
<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left: auto; margin-right: auto;">
<caption>
Sample metadata
</caption>
<thead>
<tr>
<th style="text-align:left;">
Sample
</th>
<th style="text-align:left;">
Fish ID
</th>
<th style="text-align:left;">
Batch Killed
</th>
<th style="text-align:left;">
Sex
</th>
<th style="text-align:left;">
RINe
</th>
<th style="text-align:left;">
Genotype 1
</th>
<th style="text-align:left;">
Genotype 2
</th>
<th style="text-align:left;">
Alias
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
7_KB_B12
</td>
<td style="text-align:left;">
27
</td>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD1
</td>
</tr>
<tr>
<td style="text-align:left;">
22_KB_C6
</td>
<td style="text-align:left;">
30
</td>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.6
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD2
</td>
</tr>
<tr>
<td style="text-align:left;">
16_KB_C12
</td>
<td style="text-align:left;">
40
</td>
<td style="text-align:left;">
8
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD3
</td>
</tr>
<tr>
<td style="text-align:left;">
8_KB_B3
</td>
<td style="text-align:left;">
46
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.2
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD4
</td>
</tr>
<tr>
<td style="text-align:left;">
19_KB_C3
</td>
<td style="text-align:left;">
29
</td>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.5
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD5
</td>
</tr>
<tr>
<td style="text-align:left;">
10_KB_B6
</td>
<td style="text-align:left;">
50
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD6
</td>
</tr>
<tr>
<td style="text-align:left;">
4_KB_A6
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
3
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.0
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD7
</td>
</tr>
<tr>
<td style="text-align:left;">
2_KB_A11
</td>
<td style="text-align:left;">
12
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
8.7
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI1
</td>
</tr>
<tr>
<td style="text-align:left;">
12_KB_B8
</td>
<td style="text-align:left;">
22
</td>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI2
</td>
</tr>
<tr>
<td style="text-align:left;">
15_KB_C11
</td>
<td style="text-align:left;">
45
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI3
</td>
</tr>
<tr>
<td style="text-align:left;">
18_KB_C2
</td>
<td style="text-align:left;">
47
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.2
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI4
</td>
</tr>
<tr>
<td style="text-align:left;">
9_KB_B5
</td>
<td style="text-align:left;">
21
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI5
</td>
</tr>
<tr>
<td style="text-align:left;">
6_KB_B11
</td>
<td style="text-align:left;">
26
</td>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI6
</td>
</tr>
<tr>
<td style="text-align:left;">
24_KB_C8
</td>
<td style="text-align:left;">
44
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.8
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI7
</td>
</tr>
<tr>
<td style="text-align:left;">
21_KB_C5
</td>
<td style="text-align:left;">
49
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.5
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI8
</td>
</tr>
<tr>
<td style="text-align:left;">
5_KB_A7
</td>
<td style="text-align:left;">
19
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.1
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt1
</td>
</tr>
<tr>
<td style="text-align:left;">
1_KB_A10
</td>
<td style="text-align:left;">
25
</td>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt2
</td>
</tr>
<tr>
<td style="text-align:left;">
23_KB_C7
</td>
<td style="text-align:left;">
36
</td>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.7
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt3
</td>
</tr>
<tr>
<td style="text-align:left;">
3_KB_A1
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt4
</td>
</tr>
<tr>
<td style="text-align:left;">
17_KB_C1
</td>
<td style="text-align:left;">
34
</td>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.0
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt5
</td>
</tr>
<tr>
<td style="text-align:left;">
20_KB_C4
</td>
<td style="text-align:left;">
35
</td>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt6
</td>
</tr>
<tr>
<td style="text-align:left;">
14_KB_C10
</td>
<td style="text-align:left;">
37
</td>
<td style="text-align:left;">
8
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt7
</td>
</tr>
<tr>
<td style="text-align:left;">
13_KB_B9
</td>
<td style="text-align:left;">
54
</td>
<td style="text-align:left;">
10
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt8
</td>
</tr>
<tr>
<td style="text-align:left;">
11_KB_B7
</td>
<td style="text-align:left;">
55
</td>
<td style="text-align:left;">
10
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.1
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt9
</td>
</tr>
</tbody>
</table>
<pre class="r"><code># genotypes &lt;- unique(metadata$genotype) %&gt;%
#   as.vector() %&gt;%
#   set_names(., .)
# fad &lt;- metadata$alias %&gt;%
#   str_detect(&quot;fAD&quot;)
# fai &lt;- metadata$alias %&gt;%
#   str_detect(&quot;fAI&quot;)
# wt &lt;- metadata$alias %&gt;%
#   str_detect(&quot;WT&quot;)</code></pre>
</div>
</div>
<div id="allele-specific-expression" class="section level1">
<h1>Allele specific expression</h1>
<p>Allele specific expression testing was performed at the gene-level with <code>GeneiASE</code> software. Genes were classified as showing allele specific expression (ASE) if they had a FDR-adjusted p-value &lt; 0.05.</p>
<pre class="r"><code>files &lt;- list.files(
  file.path(projDir, &quot;13_geneiase/2_ase&quot;),
  full.names = TRUE
)
samples &lt;- basename(files) %&gt;%
  str_remove(&quot;.static.pval.tsv&quot;)
topASE &lt;- lapply(files, function(x){
  sample &lt;- basename(x) %&gt;%
    str_remove(&quot;.static.pval.tsv&quot;)
  read_tsv(x) %&gt;%
    mutate(
      sig = fdr &lt; 0.05,
      sample = sample
    ) %&gt;%
    left_join(metadata[,c(&quot;sample&quot;, &quot;alias&quot;, &quot;genotype&quot;)]) %&gt;%
    dplyr::arrange(fdr) %&gt;%
    mutate(ASE = fdr &lt; 0.05)
}) %&gt;%
  set_names(metadata$alias[match(samples, metadata$sample)]) %&gt;%
  .[metadata$alias]</code></pre>
<pre class="r"><code>aseRC &lt;- readRDS(here(&quot;files/aseRC.Rds&quot;))
aseRCByGene &lt;- readRDS(here(&quot;files/aseRCByGene.Rds&quot;)) %&gt;%
  sapply(dplyr::select, gene, everything(), simplify = FALSE)</code></pre>
<pre class="r"><code>nASE &lt;- bind_rows(topASE) %&gt;%
  group_by(alias, sample, genotype) %&gt;%
  summarise(nASE = sum(ASE), total = n()) %&gt;%
  ungroup() %&gt;%
  mutate(
    alias = factor(alias),
    propASE = nASE / total
  )</code></pre>
<p>Between <em>327</em> and <em>557</em> genes were classified as showing ASE across all samples, which accounted for between <em>4.53%</em> and <em>7.14%</em> of the total genes tested.</p>
<pre class="r"><code>nASE %&gt;%
  ggplot(aes(alias, nASE, fill = genotype)) +
  geom_bar(stat = &quot;identity&quot;, colour = &quot;black&quot;) +
  scale_fill_manual(values = genoCols) +
  scale_y_continuous(breaks = seq(0, 500, 100)) +
  labs(
    x = &quot;Sample&quot;,
    y = &quot;Number of ASE genes&quot;,
    fill = &quot;Genotype&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ase_analysis_files/figure-html/nAsePlot-1.png" alt="*The number of genes classified as showing significant ASE basedon an FDR-adjusted p-value cut-off of 0.05. The number of genes showing ASE was fairly consistent across all samples and genotypes.*" width="768" />
<p class="caption">
<em>The number of genes classified as showing significant ASE basedon an FDR-adjusted p-value cut-off of 0.05. The number of genes showing ASE was fairly consistent across all samples and genotypes.</em>
</p>
</div>
<p>As the chromosomal localisation of ASE was of interest, the number of genes showing ASE was also assessed at each chromosome.</p>
<pre class="r"><code>bind_rows(topASE) %&gt;%
  left_join(chrGenes_df) %&gt;%
  mutate(chr = factor(chr, levels = str_sort(unique(chr), numeric = TRUE))) %&gt;%
  group_by(genotype, chr) %&gt;%
  summarise(nASE = sum(ASE)) %&gt;% 
  ggplot(aes(genotype, nASE, fill = genotype)) +
  geom_bar(stat = &quot;identity&quot;, colour = &quot;black&quot;) +
  facet_wrap(~chr) +
  scale_fill_manual(values = genoCols) +
  labs(
    x = &quot;Genotype&quot;,
    y = &quot;Number of ASE genes&quot;,
    fill = &quot;Genotype&quot;
  ) +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ase_analysis_files/figure-html/nAseByChrPlot-1.png" alt="*The number of genes showing ASE at each chromosome. Samples are grouped by genotype.*" width="768" />
<p class="caption">
<em>The number of genes showing ASE at each chromosome. Samples are grouped by genotype.</em>
</p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
