<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lachlan Baer" />


<title>Allele Specific Expression Analysis</title>

<script src="site_libs/header-attrs-2.9/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">210408_psen1_fADfAI_snv</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="de_analysis.html">DE analysis</a>
</li>
<li>
  <a href="de_enrichment.html">DE enrichment</a>
</li>
<li>
  <a href="ase_qc.html">ASE QC</a>
</li>
<li>
  <a href="ase_analysis.html">ASE analysis</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/baerlachlan/210408_psen1_fADfAI_snv">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Allele Specific Expression Analysis</h1>
<h4 class="author">Lachlan Baer</h4>
<h4 class="date">16 July, 2021</h4>

</div>


<div id="setup" class="section level1">
<h1>Setup</h1>
<pre class="r"><code>suppressPackageStartupMessages({
  ## Common
  library(tidyverse)
  library(magrittr)
  library(future.apply)
  library(here)
  library(AnnotationHub)
  library(purrr)
  library(scales)
  library(kableExtra)
  library(tictoc)
  library(ggrepel)
  library(RColorBrewer)
  library(ggpubr)
  library(pander)
  library(rmarkdown)
  ## Project specific
  library(statmod)
  library(fgsea)
})</code></pre>
<pre class="r"><code>if (interactive()) setwd(here::here())
theme_set(theme_bw())
cores &lt;- availableCores() - 2</code></pre>
<pre class="r"><code>source(here(&quot;R/lbFuncs.R&quot;))</code></pre>
<div id="ensdb" class="section level2">
<h2>EnsDb</h2>
<pre class="r"><code>ah &lt;- AnnotationHub() %&gt;%
  subset(species == &quot;Danio rerio&quot;) %&gt;%
  subset(rdataclass == &quot;EnsDb&quot;)
ensDb &lt;- ah[[&quot;AH83189&quot;]] ## Ens101
genes &lt;- genes(ensDb)
mcols(genes) &lt;- mcols(genes)[
  c(&quot;gene_id&quot;, &quot;gene_name&quot;, &quot;gene_biotype&quot;, &quot;entrezid&quot;)
]
exons &lt;- exonsBy(ensDb, by = &quot;gene&quot;)</code></pre>
</div>
</div>
<div id="load-data" class="section level1">
<h1>Load data</h1>
<div id="metadata" class="section level2">
<h2>Metadata</h2>
<pre class="r"><code>metadata &lt;- read_csv(here(&quot;files/samples.csv&quot;)) %&gt;%
  dplyr::select(-sample) %&gt;%
  dplyr::rename(sample = basename, genotype = Genotype) %&gt;%
  ## We need some sample aliases that follow R naming conventions
  mutate(
    alias = c(
      paste0(rep(&quot;fAD&quot;, 7), seq(1, 7)),
      paste0(rep(&quot;fAI&quot;, 8), seq(1, 8)),
      paste0(rep(&quot;wt&quot;, 9), seq(1, 9))
    )
  )
metadata$genotype &lt;- fct_relevel(
  metadata$genotype,
  c(&quot;WT&quot;, &quot;EOfAD-like/+&quot;, &quot;fAI-like/+&quot;)
)
genoCols &lt;- metadata$genotype %&gt;%
  levels() %&gt;%
  length() %&gt;%
  brewer.pal(&quot;Set1&quot;) %&gt;%
  setNames(levels(metadata$genotype))
compCols &lt;- genoCols[2:3]</code></pre>
<pre class="r"><code>metadata %&gt;%
  dplyr::rename(
    Sample = sample, `Fish ID` = fish_id, `Batch Killed` = batch_killed,
    `Genotype 1` = genotype, `Genotype 2` = Genotype_2, Alias = alias
  ) %&gt;%
  kable(
    align = &quot;l&quot;,
    caption = &quot;Sample metadata&quot;
  ) %&gt;%
  kable_styling(
    bootstrap_options = c(&quot;striped&quot;, &quot;hover&quot;, &quot;condensed&quot;, &quot;responsive&quot;)
  )</code></pre>
<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left: auto; margin-right: auto;">
<caption>
Sample metadata
</caption>
<thead>
<tr>
<th style="text-align:left;">
Sample
</th>
<th style="text-align:left;">
Fish ID
</th>
<th style="text-align:left;">
Batch Killed
</th>
<th style="text-align:left;">
Sex
</th>
<th style="text-align:left;">
RINe
</th>
<th style="text-align:left;">
Genotype 1
</th>
<th style="text-align:left;">
Genotype 2
</th>
<th style="text-align:left;">
Alias
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
7_KB_B12
</td>
<td style="text-align:left;">
27
</td>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD1
</td>
</tr>
<tr>
<td style="text-align:left;">
22_KB_C6
</td>
<td style="text-align:left;">
30
</td>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.6
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD2
</td>
</tr>
<tr>
<td style="text-align:left;">
16_KB_C12
</td>
<td style="text-align:left;">
40
</td>
<td style="text-align:left;">
8
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD3
</td>
</tr>
<tr>
<td style="text-align:left;">
8_KB_B3
</td>
<td style="text-align:left;">
46
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.2
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD4
</td>
</tr>
<tr>
<td style="text-align:left;">
19_KB_C3
</td>
<td style="text-align:left;">
29
</td>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.5
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD5
</td>
</tr>
<tr>
<td style="text-align:left;">
10_KB_B6
</td>
<td style="text-align:left;">
50
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD6
</td>
</tr>
<tr>
<td style="text-align:left;">
4_KB_A6
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
3
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.0
</td>
<td style="text-align:left;">
EOfAD-like/+
</td>
<td style="text-align:left;">
T428del/+
</td>
<td style="text-align:left;">
fAD7
</td>
</tr>
<tr>
<td style="text-align:left;">
2_KB_A11
</td>
<td style="text-align:left;">
12
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
8.7
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI1
</td>
</tr>
<tr>
<td style="text-align:left;">
12_KB_B8
</td>
<td style="text-align:left;">
22
</td>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI2
</td>
</tr>
<tr>
<td style="text-align:left;">
15_KB_C11
</td>
<td style="text-align:left;">
45
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI3
</td>
</tr>
<tr>
<td style="text-align:left;">
18_KB_C2
</td>
<td style="text-align:left;">
47
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.2
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI4
</td>
</tr>
<tr>
<td style="text-align:left;">
9_KB_B5
</td>
<td style="text-align:left;">
21
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI5
</td>
</tr>
<tr>
<td style="text-align:left;">
6_KB_B11
</td>
<td style="text-align:left;">
26
</td>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI6
</td>
</tr>
<tr>
<td style="text-align:left;">
24_KB_C8
</td>
<td style="text-align:left;">
44
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.8
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI7
</td>
</tr>
<tr>
<td style="text-align:left;">
21_KB_C5
</td>
<td style="text-align:left;">
49
</td>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.5
</td>
<td style="text-align:left;">
fAI-like/+
</td>
<td style="text-align:left;">
W233fs/+
</td>
<td style="text-align:left;">
fAI8
</td>
</tr>
<tr>
<td style="text-align:left;">
5_KB_A7
</td>
<td style="text-align:left;">
19
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.1
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt1
</td>
</tr>
<tr>
<td style="text-align:left;">
1_KB_A10
</td>
<td style="text-align:left;">
25
</td>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt2
</td>
</tr>
<tr>
<td style="text-align:left;">
23_KB_C7
</td>
<td style="text-align:left;">
36
</td>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.7
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt3
</td>
</tr>
<tr>
<td style="text-align:left;">
3_KB_A1
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
9.4
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt4
</td>
</tr>
<tr>
<td style="text-align:left;">
17_KB_C1
</td>
<td style="text-align:left;">
34
</td>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.0
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt5
</td>
</tr>
<tr>
<td style="text-align:left;">
20_KB_C4
</td>
<td style="text-align:left;">
35
</td>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt6
</td>
</tr>
<tr>
<td style="text-align:left;">
14_KB_C10
</td>
<td style="text-align:left;">
37
</td>
<td style="text-align:left;">
8
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt7
</td>
</tr>
<tr>
<td style="text-align:left;">
13_KB_B9
</td>
<td style="text-align:left;">
54
</td>
<td style="text-align:left;">
10
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.3
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt8
</td>
</tr>
<tr>
<td style="text-align:left;">
11_KB_B7
</td>
<td style="text-align:left;">
55
</td>
<td style="text-align:left;">
10
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
9.1
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
WT
</td>
<td style="text-align:left;">
wt9
</td>
</tr>
</tbody>
</table>
<pre class="r"><code>genotypes &lt;- unique(metadata$genotype) %&gt;%
  as.vector() %&gt;%
  set_names(., .)
fad &lt;- metadata$alias %&gt;%
  str_detect(&quot;fAD&quot;)
fai &lt;- metadata$alias %&gt;%
  str_detect(&quot;fAI&quot;)
wt &lt;- metadata$alias %&gt;%
  str_detect(&quot;WT&quot;)</code></pre>
</div>
<div id="differential-expression" class="section level2">
<h2>Differential Expression</h2>
<pre class="r"><code>topDE &lt;- readRDS(here(&quot;files/toptable_raw.rds&quot;)) %&gt;%
  set_names(c(&quot;fAD&quot;, &quot;fAI&quot;))</code></pre>
</div>
<div id="geneiase" class="section level2">
<h2>GeneiASE</h2>
<pre class="r"><code>files &lt;- list.files(
  &quot;/hpcfs/users/a1647910/210408_psen1_fADfAI_snv/12_geneiase/ase&quot;,
  full.names = TRUE
)
samples &lt;- basename(files) %&gt;%
  str_remove(&quot;.static.pval.tsv&quot;)
drChrs &lt;- seq(1:25)</code></pre>
<pre class="r"><code>topASE &lt;- lapply(files, function(x){
  sample &lt;- basename(x) %&gt;%
    str_remove(&quot;.static.pval.tsv&quot;)
  read_tsv(x) %&gt;%
    mutate(
      sig = fdr &lt; 0.05,
      sample = sample
    ) %&gt;%
    left_join(metadata[,c(&quot;sample&quot;, &quot;alias&quot;, &quot;genotype&quot;)]) %&gt;%
    dplyr::arrange(fdr) %&gt;%
    mutate(ASE = fdr &lt; 0.05)
}) %&gt;%
  set_names(metadata$alias[match(samples, metadata$sample)]) %&gt;%
  .[metadata$alias]</code></pre>
<pre class="r"><code>nASE &lt;- sapply(topASE, function(x){
  dplyr::filter(x, ASE) %&gt;%
    nrow()
})</code></pre>
</div>
</div>
<div id="enrichment-testing" class="section level1">
<h1>Enrichment testing</h1>
<div id="fishers-exact-test" class="section level2">
<h2>Fisher’s exact test</h2>
<p>The first and most simple approach to testing for overall enrichment of DE genes showing ASE is Fisher’s exact test. 2x2 contingency tables were constructed for each mutant genotype (fAD and fAI). Genotype fAD showed 13 DE genes in comparison to WT, while genes presenting static ASE ranged from 850 to 1045 across 24 samples. Genotype fAI showed 5 DE genes in comparison to WT, while genes presenting static ASE ranged from 976 to 1180 across 24 samples.</p>
<pre class="r"><code>fisherTabs &lt;- topASE[str_detect(names(topASE), (&quot;fAD|fAI&quot;))] %&gt;%
  lapply(function(x){
    geno &lt;- unique(x$alias)
    if (str_detect(geno, &quot;fAD&quot;)) {
      dat &lt;- topDE$fAD[,c(&quot;gene_id&quot;, &quot;DE&quot;)] %&gt;%
        left_join(x[,c(&quot;feat&quot;, &quot;ASE&quot;)], by = c(&quot;gene_id&quot; = &quot;feat&quot;)) %&gt;%
        mutate(
          DE = ifelse(DE == TRUE, &quot;DE&quot;, &quot;Not DE&quot;),
          ASE = ifelse(ASE == TRUE, &quot;ASE&quot;, &quot;No ASE&quot;),
          ASE = ifelse(is.na(ASE), &quot;No ASE&quot;, ASE)
        )
      tab &lt;- table(dat$DE, dat$ASE)
      return(tab)
    }
    if (str_detect(geno, &quot;fAI&quot;)) {
      dat &lt;- topDE$fAI[,c(&quot;gene_id&quot;, &quot;DE&quot;)] %&gt;%
        left_join(x[,c(&quot;feat&quot;, &quot;ASE&quot;)], by = c(&quot;gene_id&quot; = &quot;feat&quot;)) %&gt;%
        mutate(
          DE = ifelse(DE == TRUE, &quot;DE&quot;, &quot;Not DE&quot;),
          ASE = ifelse(ASE == TRUE, &quot;ASE&quot;, &quot;No ASE&quot;),
          ASE = ifelse(is.na(ASE), &quot;No ASE&quot;, ASE)
        )
      tab &lt;- table(dat$DE, dat$ASE)
      return(tab)
    }
  })</code></pre>
<p>Concerns with statistical power for Fisher’s exact test were assessed due to the small number of DE genes for both genotypes. This was performed with the <code>power.fisher.test()</code> function using 1000 simulations for each sample.</p>
<pre class="r"><code>set.seed(17)
fisherSim &lt;- sapply(fisherTabs, function(x){
  p1 &lt;- x[1,1] / (x[2,1])
  n1 &lt;- (x[1,1] + x[2,1])
  p2 &lt;- x[1,2] / (x[2,2])
  n2 &lt;- (x[1,2] + x[2,2])
  power.fisher.test(p1, p2, n1, n2, alpha = 0.05, nsim = 1000)
})</code></pre>
<pre class="r"><code>fisherSim %&gt;%
  enframe(name = &quot;Sample&quot;, value = &quot;Power&quot;) %&gt;%
  left_join(metadata, by = c(&quot;Sample&quot; = &quot;alias&quot;)) %&gt;%
  ggplot(aes(Sample, Power)) +
  geom_bar(aes(fill = genotype), stat = &quot;identity&quot;) +
  scale_fill_manual(values = compCols) +
  labs(fill = &quot;Genotype&quot;)</code></pre>
<p><img src="ase_analysis_files/figure-html/unnamed-chunk-12-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Fisher’s exact test lacks power to detect enrichment of DE genes presenting ASE. The test was stil performed as an initial insight, but other methods must be considered to make proper conclusions.</p>
<pre class="r"><code>fisherTest &lt;- lapply(fisherTabs, fisher.test)
fisherP &lt;- sapply(fisherTest, function(x){x$p.value})</code></pre>
<pre class="r"><code>fisherP %&gt;%
  enframe(name = &quot;Sample&quot;, value = &quot;p&quot;) %&gt;%
  left_join(metadata, by = c(&quot;Sample&quot; = &quot;alias&quot;)) %&gt;%
  ggplot(aes(Sample, -log10(p))) +
  geom_bar(aes(fill = genotype), stat = &quot;identity&quot;) +
  geom_hline(yintercept = -log10(0.05), linetype = &quot;dashed&quot;) +
  scale_fill_manual(values = compCols) +
  labs(fill = &quot;Genotype&quot;, y = log10plab)</code></pre>
<p><img src="ase_analysis_files/figure-html/unnamed-chunk-14-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="gene-set-enrichment-analysis" class="section level2">
<h2>Gene Set Enrichment Analysis</h2>
<div id="setup-1" class="section level3">
<h3>Setup</h3>
<pre class="r"><code>chrGenes &lt;- genes %&gt;%
  as_tibble() %&gt;%
  dplyr::filter(seqnames %in% drChrs) %&gt;%
  droplevels() %&gt;%
  split(.$seqnames) %&gt;%
  lapply(pull, gene_id) %&gt;%
  set_names(paste0(&quot;chr&quot;, names(.)))</code></pre>
<pre class="r"><code>wilsonConf &lt;- function(pvals, conf.level = 0.95){
  p &lt;- mean(pvals)
  n &lt;- length(pvals)
  alpha &lt;- 1 - conf.level
  alpha2 &lt;- 0.5 * alpha
  z &lt;- qnorm(1 - alpha2)
  z2 &lt;- z * z
  p1 &lt;- p + 0.5 * z2/n
  p2 &lt;- z * sqrt((p * (1 - p) + 0.25 * z2/n)/n)
  p3 &lt;- 1 + z2/n
  lcl &lt;- (p1 - p2)/p3
  ucl &lt;- (p1 + p2)/p3
  res.wilson &lt;- tibble(
    mean = p,
    conf = p2,
    lower = lcl,
    upper = ucl,
    sig = mean &lt; 0.05
  )
  return(res.wilson)
}</code></pre>
<pre class="r"><code>intSizeByN &lt;- lapply(seq(1:1000), function(n){
  p &lt;- 0.05
  alpha &lt;- 1 - p
  alpha2 &lt;- 0.5 * alpha
  z &lt;- qnorm(1 - alpha2)
  z2 &lt;- z * z
  p1 &lt;- p + 0.5 * z2/n
  p2 &lt;- z * sqrt((p * (1 - p) + 0.25 * z2/n)/n)
  p3 &lt;- 1 + z2/n
  lcl &lt;- (p1 - p2)/p3
  ucl &lt;- (p1 + p2)/p3
  res.wilson &lt;- tibble(
    p = p,
    n = n,
    lower = lcl,
    upper = ucl
  )
  return(res.wilson)
})
intSizeByN %&gt;%
  bind_rows() %&gt;%
  mutate(size = upper - lower) %&gt;%
  ggplot(aes(n, size)) +
  geom_line() +
  labs(y = &quot;Wilson interval size&quot;) +
  scale_y_continuous(breaks = seq(0, 0.03, by = 0.001)) +
  scale_x_continuous(breaks = seq(0, 1000, by = 100))</code></pre>
<p><img src="ase_analysis_files/figure-html/unnamed-chunk-17-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="non-directional" class="section level3">
<h3>Non-directional</h3>
<pre class="r"><code>permRanks_nd &lt;- function(topASE){
  topASE[,c(&quot;feat&quot;, &quot;fdr&quot;)] %&gt;%
    mutate(
      fdr = ifelse(fdr == 0, 0.000001, fdr),
      stat = -log10(fdr)
    ) %&gt;%
    split(f = .$stat) %&gt;%
    lapply(function(x){
      pull(x, feat) %&gt;%
        sample(length(.), replace = FALSE)
    }) %&gt;%
    unlist(use.names = FALSE) %&gt;%
    enframe(name = NULL, value = &quot;geneid&quot;) %&gt;%
    mutate(stat = seq(1:nrow(.))) %&gt;%
    with(structure(stat, names = geneid)) %&gt;%
    rev()
}</code></pre>
<pre class="r"><code>perms &lt;- 500
file_nd &lt;- paste0(
  &quot;/hpcfs/users/a1647910/210408_psen1_fADfAI_snv&quot;,
  &quot;/files/fgsea_nd_&quot;,
  perms,
  &quot;perms.Rds&quot;
)</code></pre>
<pre class="r"><code>## This code chunk has a lengthy run-time and will not run if the object has already been saved
## Due to its large size the object remains on the HPC system
## Across 6 threads it takes 1-2 hours to complete
if (!file.exists(file_nd)) {
  plan(list(tweak(multisession, workers = cores), sequential))
  
  ranks_nd &lt;- future_sapply(1:perms, function(...){
    future_sapply(topASE, permRanks_nd, future.seed = 17)
  }, future.seed = 17) %&gt;%
    t() %&gt;%
    as_tibble()
  fgsea_nd &lt;- future_lapply(as.list(ranks_nd), function(x){
    future_lapply(x, function(y){
      fgseaMultilevel(chrGenes, y, scoreType = &quot;pos&quot;, eps = 0)
    }, future.seed = 17)
  }, future.seed = 17) %&gt;%
    set_names(colnames(ranks_nd))
  saveRDS(fgsea_nd, file_nd)
  
  plan(sequential)
}</code></pre>
<pre class="r"><code>fgsea_nd &lt;- readRDS(file_nd)</code></pre>
<pre class="r"><code>pvals_nd &lt;- lapply(fgsea_nd, function(x){
  lapply(seq_along(x), function(y){
    x[[y]][,c(&quot;pathway&quot;, &quot;padj&quot;)] %&gt;%
      dplyr::rename(!!as.character(y) := padj)
  }) %&gt;%
    purrr::reduce(left_join) %&gt;%
    column_to_rownames(&quot;pathway&quot;) %&gt;%
    t() %&gt;%
    as_tibble()
})</code></pre>
<pre class="r"><code>mean_nd &lt;- lapply(pvals_nd, function(sample){
  chrList &lt;- as.list(sample)
  lapply(seq_along(chrList), function(x){
    chr &lt;- names(chrList)[x]
    wilsonConf(chrList[[x]]) %&gt;%
      mutate(chromosome = as.factor(chr))
  }) %&gt;%
    bind_rows() %&gt;%
    dplyr::select(chromosome, everything())
}) </code></pre>
<pre class="r"><code>lapply(seq_along(mean_nd), function(ind){
  x &lt;- mean_nd[[ind]]
  x %&gt;%
    mutate(
      sample = names(mean_nd)[ind],
      chromosome = fct_relevel(chromosome, str_sort(chromosome, numeric = TRUE))
    )
}) %&gt;%
  bind_rows() %&gt;%
  left_join(metadata[,c(&quot;alias&quot;, &quot;genotype&quot;)], by = c(&quot;sample&quot; = &quot;alias&quot;)) %&gt;%
  ggplot(aes(sample, -log10(mean))) +
  geom_bar(stat = &quot;identity&quot;, aes(fill = genotype)) +
  geom_hline(yintercept = -log10(0.05), linetype = &quot;dashed&quot;) +
  facet_wrap(~ chromosome) +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    legend.position = &quot;bottom&quot;
  ) +
  scale_fill_manual(values = genoCols) +
  labs(y = log10plab)</code></pre>
<p><img src="ase_analysis_files/figure-html/unnamed-chunk-24-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="directional" class="section level3">
<h3>Directional</h3>
<pre class="r"><code>permRanks_d &lt;- function(topASE){
  geno &lt;- unique(topASE$alias)
  if (str_detect(geno, &quot;fAD&quot;)) {
    dat &lt;- topASE[,c(&quot;feat&quot;, &quot;fdr&quot;)] %&gt;%
      left_join(
        topDE$fAD[,c(&quot;gene_id&quot;, &quot;logFC&quot;)],
        by = c(&quot;feat&quot; = &quot;gene_id&quot;)
      )
  }
  if (str_detect(geno, &quot;fAI&quot;)) {
    dat &lt;- topASE[,c(&quot;feat&quot;, &quot;fdr&quot;)] %&gt;%
      left_join(
        topDE$fAI[,c(&quot;gene_id&quot;, &quot;logFC&quot;)],
        by = c(&quot;feat&quot; = &quot;gene_id&quot;)
      )
  }
  dat %&gt;%
    mutate(
      fdr = ifelse(fdr == 0, 0.000001, fdr),
      stat = sign(logFC) * -log10(fdr)
    ) %&gt;%
    dplyr::arrange(stat) %&gt;%
    split(f = .$stat) %&gt;%
    lapply(function(x){
      pull(x, feat) %&gt;%
        sample(length(.), replace = FALSE)
    }) %&gt;%
    unlist(use.names = FALSE) %&gt;%
    enframe(name = NULL, value = &quot;geneid&quot;) %&gt;%
    mutate(stat = seq(1:nrow(.))) %&gt;%
    with(structure(stat, names = geneid)) %&gt;%
    rev()
}</code></pre>
<pre class="r"><code>perms &lt;- 500
file_d &lt;- paste0(
  &quot;/hpcfs/users/a1647910/210408_psen1_fADfAI_snv&quot;,
  &quot;/files/fgsea_d_&quot;,
  perms,
  &quot;perms.Rds&quot;
)</code></pre>
<pre class="r"><code>## This code chunk has a lengthy run-time and will not run if the object has already been saved
## Due to its large size the object remains on the HPC system
## Across 6 threads it takes 1-2 hours to complete
if (!file.exists(file_d)) {
  plan(list(tweak(multisession, workers = 6), sequential))
  
  ranks_d &lt;- future_sapply(1:perms, function(...){
    future_sapply(
      topASE[str_detect(names(topASE), &quot;(fAD|fAI)&quot;)],
      permRanks_d,
      future.seed = 17
    )
  }, future.seed = 17) %&gt;%
    t() %&gt;%
    as_tibble()
  fgsea_d &lt;- future_lapply(as.list(ranks_d), function(x){
    future_lapply(x, function(y){
      fgseaMultilevel(chrGenes, y, scoreType = &quot;pos&quot;, eps = 0)
    }, future.seed = 17)
  }, future.seed = 17) %&gt;%
    set_names(colnames(ranks_d))
  saveRDS(fgsea_d, file_d)
  
  plan(sequential)
}</code></pre>
<pre class="r"><code>fgsea_d &lt;- readRDS(file_d)</code></pre>
<pre class="r"><code>pvals_d &lt;- lapply(fgsea_d, function(x){
  lapply(seq_along(x), function(y){
    x[[y]][,c(&quot;pathway&quot;, &quot;padj&quot;)] %&gt;%
      dplyr::rename(!!as.character(y) := padj)
  }) %&gt;%
    purrr::reduce(left_join) %&gt;%
    column_to_rownames(&quot;pathway&quot;) %&gt;%
    t() %&gt;%
    as_tibble()
})</code></pre>
<pre class="r"><code>mean_d &lt;- lapply(pvals_d, function(sample){
  chrList &lt;- as.list(sample)
  lapply(seq_along(chrList), function(x){
    chr &lt;- names(chrList)[x]
    wilsonConf(chrList[[x]]) %&gt;%
      mutate(chromosome = as.factor(chr))
  }) %&gt;%
    bind_rows() %&gt;%
    dplyr::select(chromosome, everything())
}) </code></pre>
<pre class="r"><code>lapply(seq_along(mean_d), function(ind){
  x &lt;- mean_d[[ind]]
  x %&gt;%
    mutate(
      sample = names(mean_d)[ind],
      chromosome = fct_relevel(chromosome, str_sort(chromosome, numeric = TRUE))
    )
}) %&gt;%
  bind_rows() %&gt;%
  left_join(metadata[,c(&quot;alias&quot;, &quot;genotype&quot;)], by = c(&quot;sample&quot; = &quot;alias&quot;)) %&gt;%
  ggplot(aes(sample, -log10(mean))) +
  geom_bar(stat = &quot;identity&quot;, aes(fill = genotype)) +
  geom_hline(yintercept = -log10(0.05), linetype = &quot;dashed&quot;) +
  facet_wrap(~ chromosome) +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    legend.position = &quot;bottom&quot;
  ) +
  scale_fill_manual(values = compCols) +
  labs(y = log10plab)</code></pre>
<p><img src="ase_analysis_files/figure-html/unnamed-chunk-31-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
</div>
</div>
<div id="tmp" class="section level1">
<h1>tmp</h1>
<pre class="r"><code>chrGenes_df &lt;- chrGenes %&gt;%
  lapply(list) %&gt;%
  as_tibble() %&gt;%
  pivot_longer(cols = everything(), names_to = &quot;chr&quot;, values_to= &quot;feat&quot;) %&gt;%
  unnest(feat)</code></pre>
<pre class="r"><code>## Give number of samples showing ASE for each gene
aseByGene &lt;- lapply(topASE, dplyr::filter, ASE) %&gt;%
  bind_rows() %&gt;%
  mutate(genoAlias = case_when(
    str_detect(alias, &quot;fAD&quot;) ~ &quot;fAD&quot;,
    str_detect(alias, &quot;fAI&quot;) ~ &quot;fAI&quot;,
    str_detect(alias, &quot;wt&quot;) ~ &quot;WT&quot;,
  )) %&gt;%
  group_by(feat, genoAlias) %&gt;%
  tally() %&gt;%
  ungroup() %&gt;%
  pivot_wider(names_from = genoAlias, values_from = n, values_fill = 0) %&gt;%
  left_join(chrGenes_df)</code></pre>
<pre class="r"><code>## Gives number of genes showingASE in at least 1 sample
aseByChr &lt;- sapply(c(&quot;fAD&quot;, &quot;fAI&quot;, &quot;WT&quot;), function(geno){
  aseByGene %&gt;%
    dplyr::filter(!!sym(geno) != 0) %&gt;%
    group_by(chr) %&gt;%
    tally(name = geno) %&gt;%
    arrange(desc(!!sym(geno)))
}, simplify = FALSE)</code></pre>
<pre class="r"><code>noAseByGene &lt;- lapply(topASE, dplyr::filter, !ASE) %&gt;%
  bind_rows() %&gt;%
  mutate(genoAlias = case_when(
    str_detect(alias, &quot;fAD&quot;) ~ &quot;fAD&quot;,
    str_detect(alias, &quot;fAI&quot;) ~ &quot;fAI&quot;,
    str_detect(alias, &quot;wt&quot;) ~ &quot;WT&quot;,
  )) %&gt;%
  group_by(feat, genoAlias) %&gt;%
  tally() %&gt;%
  ungroup() %&gt;%
  pivot_wider(names_from = genoAlias, values_from = n, values_fill = 0) %&gt;%
  left_join(chrGenes_df)</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
