---
title: "Variants Analysis"
subtitle: "210408_psen1_fADfAI dataset"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    theme: flatly
    toc: true
    toc_float: true
    toc_depth: 4
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options:
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE,
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Publication plots

```{r}
# get_trackList <- function(
    #     chr, genes = GRanges(), mutation = GRanges(),
#     dist, dist_type = "l", dist_trans = 1, highlight_genes = TRUE,
#     lfc = NULL, lfc_type = "heatmap", lfc_trans = 1/3,
#     title
# ){
#   tracks <- list()
#   
#   if (length(mutation)) {
#     mutation <- mutation %>%
#       plyranges::filter(seqnames == chr) %>%
#       plyranges::mutate(symbol = gene_name)
#     mutant_track <- GeneRegionTrack(
#       range = mutation,
#       transcriptAnnotation = "symbol",
#       col = "white",
#       fill = "white",
#       showTranscriptId = TRUE,
#       fontcolor.group = "red",
#       cex.group = 0.6,
#       size = 0.3
#     )
#     tracks <- append(tracks, mutant_track)
#   }
#   
#   axis_track <- GenomeAxisTrack(
#     add53 = TRUE,
#     add35 = TRUE,
#     name = paste0("Chr", chr),
#     showTitle = TRUE
#   )
#   if (length(mutation)) {
#     axis_track <- HighlightTrack(
#       trackList = list(axis_track),
#       range = mutation
#     )
#   }
#   tracks <- append(tracks, axis_track)
#   
#   if (length(genes)) {
#     genes <- genes %>%
#       plyranges::filter(seqnames == chr) %>%
#       plyranges::mutate(symbol = gene_name)
#     gene_track <- GeneRegionTrack(
#       range = genes,
#       name = "DE",
#       transcriptAnnotation = "symbol",
#       showTranscriptId = TRUE,
#       fontcolor.group = 1,
#       cex.group = 0.5,
#       size = 0.3
#     )
#     tracks <- append(tracks, gene_track)
#   }
#   
#   dist_track <- dist[,"diversity"] %>%
#     plyranges::filter(seqnames == chr) %>%
#     DataTrack(
#       type = dist_type,
#       name = "Between-group genetic dissimilarity",
#       window = -1,
#       windowSize = 1,
#       col = "grey30",
#       col.axis = "black",
#       yTicksAt = seq(0, 1, 0.1),
#       ylim = c(0, 1),
#       transformation = function(x){x^(dist_trans)}
#     )
#   if (highlight_genes) {
#     if (all(length(genes) & length(mutation))) {
#       ranges <- c(genes, mutation)
#     } else if (length(genes)) {
#       ranges <- genes
#     } else if (length(mutation)) {
#       ranges <- mutation
#     }
#     dist_track <- HighlightTrack(
#       trackList = list(dist_track),
#       range = c(genes, mutation),
#       col = c(rep("#ffd1d1", length(genes)), rep("red", length(mutation))),
#       fill = c(rep("#ffd1d1", length(genes)), rep("red", length(mutation)))
#     )
#   }
#   tracks <- append(tracks, dist_track)
#   
#   if (length(lfc)) {
#     lfc_track <- lfc[,"lfc_rollmean"] %>%
#       plyranges::filter(seqnames == chr) %>%
#       DataTrack(
#         type = lfc_type,
#         name = paste0("abs.logFC^(1/3)"),
#         size = 1.5,
#         window = -1,
#         windowSize = 1,
#         col.axis = "black",
#         showColorBar = TRUE,
#         gradient = viridis(100, option = "D"),
#         transformation = function(x){x^(lfc_trans)}
#       )
#     tracks <- append(tracks, lfc_track)
#   }
#   
#   tracks
# }
```

```{r}
# chr <- "17"
# ind <- 1
# a <- as.ggplot(
#   ~plotTracks(
#     get_trackList(
#       chr = chr,
#       genes = genes[deGenes[[ind]]$gene_id],
#       mutation = genes[genes$gene_name == "psen1"],
#       dist = allele_dist[[ind]], dist_type = "l", dist_trans = 1,
#       lfc = lfc[[ind]], lfc_type = "heatmap", lfc_trans = 1/3,
#       title = ""
#     ),
#     cex.main = 0.5,
#     col.title = "black",
#     background.title = "white"
#   )
# )
```

```{r}
# ggsave(
#   "~/phd/publications/fisher/figures/figure_5.png",
#   device = "png",
#   width = 18,
#   height = 6,
#   units = "cm",
#   plot = as.grob(a)
# )
```

```{r}
plot_for_pub <- function(
    chr, genes = GRanges(), mutation = GRanges(),
    dist, dist_type = "l", dist_trans = 1, highlight_genes = TRUE,
    lfc = NULL, lfc_type = "heatmap", lfc_trans = 1/3,
    title
){
  tracks <- list()
  
  if (length(mutation)) {
    mutation <- mutation %>%
      plyranges::filter(seqnames == chr) %>%
      plyranges::mutate(symbol = gene_name)
    mutant_track <- GeneRegionTrack(
      range = mutation,
      transcriptAnnotation = "symbol",
      col = "white",
      fill = "white",
      showTranscriptId = TRUE,
      fontcolor.group = "red",
      cex.group = 0.6,
      size = 0.5,
      name = NULL
    )
    tracks <- append(tracks, mutant_track)
  }
  
  axis_track <- GenomeAxisTrack(
    add53 = TRUE,
    add35 = TRUE,
    name = paste0("Chr", chr),
    showTitle = TRUE,
    size = 1
  )
  if (length(mutation)) {
    axis_track <- HighlightTrack(
      trackList = list(axis_track),
      range = mutation
    )
  }
  tracks <- append(tracks, axis_track)
  
  if (length(genes)) {
    genes <- genes %>%
      plyranges::filter(seqnames == chr) %>%
      plyranges::mutate(symbol = gene_name)
    lapply(1:length(genes), function(i){
      mid <- floor(length(genes) / 2)
      if (i == mid) {
        name <- "DE"
      } else {
        name <- ""
      }
      gene_track <- GeneRegionTrack(
        range = genes[i,],
        name = name,
        geneSymbols = TRUE,
        fontcolor.group = 1,
        cex.group = 0.5,
        size = 2 / length(genes)
      )
      tracks <<- append(tracks, gene_track)
    })
    # gene_track <- GeneRegionTrack(
    #   range = genes,
    #   name = "DE",
    #   geneSymbols = TRUE,
    #   fontcolor.group = 1,
    #   cex.group = 0.5,
    #   size = 1
    # )
    # tracks <- append(tracks, gene_track)
  }
  
  dist_track <- dist[,"diversity"] %>%
    plyranges::filter(seqnames == chr) %>%
    DataTrack(
      type = dist_type,
      name = "DAR",
      size = 8,
      window = -1,
      windowSize = 1,
      col = "grey30",
      col.axis = "black",
      yTicksAt = seq(0, 1, 0.1),
      ylim = c(0, 1),
      transformation = function(x){x^(dist_trans)}
    )
  if (highlight_genes) {
    if (all(length(genes) & length(mutation))) {
      ranges <- c(genes, mutation)
    } else if (length(genes)) {
      ranges <- genes
    } else if (length(mutation)) {
      ranges <- mutation
    }
    dist_track <- HighlightTrack(
      trackList = list(dist_track),
      range = c(genes, mutation),
      col = c(rep("#ffd1d1", length(genes)), rep("red", length(mutation))),
      fill = c(rep("#ffd1d1", length(genes)), rep("red", length(mutation)))
    )
  }
  tracks <- append(tracks, dist_track)
  
  if (length(lfc)) {
    lfc_track <- lfc[,"lfc_rollmean"] %>%
      plyranges::filter(seqnames == chr) %>%
      DataTrack(
        type = lfc_type,
        name = paste0("abs.logFC^(1/3)"),
        size = 2,
        window = -1,
        windowSize = 1,
        col.axis = "black",
        showColorBar = TRUE,
        gradient = viridis(100, option = "D"),
        transformation = function(x){x^(lfc_trans)}
      )
    tracks <- append(tracks, lfc_track)
  }
  
  plotTracks(
    trackList = tracks,
    main = title,
    cex.main = 1,
    cex.title = 0.6,
    col.title = "black",
    background.title = "white"
  )
}
```

```{r}
png(
  filename = "~/phd/publications/fisher/figures/figure_5.png",
  width = 32,
  height = 16,
  units = "cm",
  res = 300
)
```

```{r}
chr <- "17"
ind <- 1
plot_for_pub(
  chr = chr,
  genes = genes[deGenes[[ind]]$gene_id],
  mutation = genes[genes$gene_name == "psen1"],
  dist = allele_dist[[ind]], dist_type = "l", dist_trans = 1,
  lfc = lfc[[ind]], lfc_type = "heatmap", lfc_trans = 1/3,
  title = ""
)
```

```{r}
dev.off()
```

```{r}
allele_dist[[2]] %>%
  split(seqnames(.)) %>%
  lapply(function(chr){
    tibble(
      chromosome = as.vector(seqnames(chr)),
      div = chr$diversity
    )
  }) %>%
  purrr::reduce(rbind) %>%
  mutate(
    chromosome = factor(
      chromosome,
      # Draw mutant chr on top
      levels = unique(c(
        chromosome[chromosome != "7"],
        chromosome[chromosome == "7"]
      ))
    ),
    mutant_chromosome = chromosome == "7"
  ) %>%
  ggplot(aes(x = div, group = chromosome, colour = mutant_chromosome)) +
  stat_ecdf() +
  scale_colour_manual(values = c("grey50", "red")) +
  labs(
    title = names(allele_dist[2]),
    x = "DAR",
    y = "F(DAR)"
  )
```

```{r}
ggsave(
  "~/phd/publications/fisher_pub/figures/supp_3.png",
  device = "png",
  width = 18,
  height = 8,
  units = "cm"
)
```