---
title: "Allele Specific Expression Analysis"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    theme: flatly
    toc: true
    toc_float: true
    toc_depth: 4
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE, 
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Setup 

```{r packages}
suppressPackageStartupMessages({
  ## Common
  library(tidyverse)
  library(magrittr)
  library(future.apply)
  library(here)
  library(AnnotationHub)
  library(purrr)
  library(scales)
  library(kableExtra)
  library(tictoc)
  library(ggrepel)
  library(RColorBrewer)
  library(ggpubr)
  library(pander)
  library(rmarkdown)
  ## Project specific
  library(statmod)
  library(fgsea)
})
```

```{r options}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
cores <- availableCores() - 2
```

```{r lbFuncs}
source("~/bioinformatics/bioToolkit/lbFuncs.R")
```

```{r projDir}
## Select base directory for the project
projDir <- here()
# projDir <- "/hpcfs/users/a1647910/210408_psen1_fADfAI_snv"
```

## EnsDb

```{r ensDb}
ah <- AnnotationHub() %>%
  subset(species == "Danio rerio") %>%
  subset(rdataclass == "EnsDb")
ensDb <- ah[["AH83189"]] ## Ens101
genes <- genes(ensDb)
mcols(genes) <- mcols(genes)[
  c("gene_id", "gene_name", "gene_biotype", "entrezid")
]
exons <- exonsBy(ensDb, by = "gene")
drChrs <- seq(1:25)
```

```{r}
chrGenes <- genes %>%
  as_tibble() %>%
  dplyr::filter(seqnames %in% drChrs) %>%
  droplevels() %>%
  split(.$seqnames) %>%
  lapply(pull, gene_id) %>%
  set_names(paste0("chr", names(.)))
```

```{r}
chrGenes_df <- chrGenes %>%
  lapply(list) %>%
  as_tibble() %>%
  pivot_longer(cols = everything(), names_to = "chr", values_to= "feat") %>%
  unnest(feat)
```

## Metadata

```{r metadata}
metadata <- read_csv(file.path(projDir, "files/samples.csv")) %>%
  dplyr::select(-sample) %>%
  dplyr::rename(sample = basename, genotype = Genotype) %>%
  ## We need some sample aliases that follow R naming conventions
  mutate(
    alias = c(
      paste0(rep("fAD", 7), seq(1, 7)),
      paste0(rep("fAI", 8), seq(1, 8)),
      paste0(rep("wt", 9), seq(1, 9))
    )
  )
metadata$genotype <- fct_relevel(
  metadata$genotype,
  c("EOfAD-like/+", "fAI-like/+", "WT")
)
genoCols <- metadata$genotype %>%
  levels() %>%
  length() %>%
  brewer.pal("Set1") %>%
  setNames(levels(metadata$genotype))
compCols <- genoCols[1:2]
```

```{r showMetadata}
metadata %>%
  dplyr::rename(
    Sample = sample, `Fish ID` = fish_id, `Batch Killed` = batch_killed,
    `Genotype 1` = genotype, `Genotype 2` = Genotype_2, Alias = alias
  ) %>%
  kable(
    align = "l",
    caption = "Sample metadata"
  ) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive")
  )
```

```{r genotypes}
genotypes <- unique(metadata$genotype) %>%
  as.vector() %>%
  set_names(., .)
fad <- metadata$alias %>%
  str_detect("fAD")
fai <- metadata$alias %>%
  str_detect("fAI")
wt <- metadata$alias %>%
  str_detect("WT")
```

# Allele specific expression

Allele specific expression testing was performed at the gene-level with `GeneiASE` software.
Genes were classified as showing allele specific expression (ASE) if they had a FDR-adjusted p-value < 0.05.

```{r aseData}
files <- list.files(
  file.path(projDir, "13_geneiase/2_ase"),
  full.names = TRUE
)
samples <- basename(files) %>%
  str_remove(".static.pval.tsv")
topASE <- lapply(files, function(x){
  sample <- basename(x) %>%
    str_remove(".static.pval.tsv")
  read_tsv(x) %>%
    mutate(
      sig = fdr < 0.05,
      sample = sample
    ) %>%
    left_join(metadata[,c("sample", "alias", "genotype")]) %>%
    dplyr::arrange(fdr) %>%
    mutate(ASE = fdr < 0.05)
}) %>%
  set_names(metadata$alias[match(samples, metadata$sample)]) %>%
  .[metadata$alias]
```

```{r aseCounts}
aseRC <- readRDS(here("files/aseRC.Rds"))
aseRCByGene <- readRDS(here("files/aseRCByGene.Rds")) %>%
  sapply(dplyr::select, gene, everything(), simplify = FALSE)
```

```{r nAse}
nASE <- bind_rows(topASE) %>%
  group_by(alias, sample, genotype) %>%
  summarise(nASE = sum(ASE), total = n()) %>%
  ungroup() %>%
  mutate(
    alias = factor(alias),
    propASE = nASE / total
  )
```

Between `r pander(range(nASE$nASE))` genes were classified as showing ASE across all samples, which accounted for between `r pander(range(percent_format(0.01)(nASE$propASE)))` of the total genes tested.

```{r nAsePlot, fig.cap="*The number of genes classified as showing significant ASE basedon an FDR-adjusted p-value cut-off of 0.05. The number of genes showing ASE was fairly consistent across all samples and genotypes.*"}
nASE %>%
  ggplot(aes(alias, nASE, fill = genotype)) +
  geom_bar(stat = "identity", colour = "black") +
  scale_fill_manual(values = genoCols) +
  scale_y_continuous(breaks = seq(0, 500, 100)) +
  labs(
    x = "Sample",
    y = "Number of ASE genes",
    fill = "Genotype"
  )
```

As the chromosomal localisation of ASE was of interest, the number of genes showing ASE was also assessed at each chromosome.

```{r nAseByChrPlot, fig.height=10, fig.cap="*The number of genes showing ASE at each chromosome. Samples are grouped by genotype.*"}
bind_rows(topASE) %>%
  left_join(chrGenes_df) %>%
  mutate(chr = factor(chr, levels = str_sort(unique(chr), numeric = TRUE))) %>%
  group_by(genotype, chr) %>%
  summarise(nASE = sum(ASE)) %>% 
  ggplot(aes(genotype, nASE, fill = genotype)) +
  geom_bar(stat = "identity", colour = "black") +
  facet_wrap(~chr) +
  scale_fill_manual(values = genoCols) +
  labs(
    x = "Genotype",
    y = "Number of ASE genes",
    fill = "Genotype"
  ) +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank(),
    legend.position = "bottom"
  )
```

# ASE overlap with DE genes

```{r deData}
topDE <- readRDS(file.path(projDir, "files/topTables.Rds")) %>%
  set_names(c("fAD", "fAI"))
```

In order to make any predictions regarding the possibility for ASE to affect DE results, the DE genes were assessed to determine which category they fell into:

1. Included in testing for ASE
2. Filtered by QC measures
3. Not detected (no variants)

As ASE testing was performed using a static (within sample) approach, the overlaps were also assessed on a per sample basis.

```{r deGenes}
deGenes <- sapply(topDE, dplyr::filter, DE, simplify = FALSE)
```

```{r}
sapply(deGenes, function(x){
  lapply(aseRCByGene, function(y){
    dplyr::filter(y, gene %in% x$gene_id)
  }) %>%
    bind_rows() %>%
    group_by(gene, sample) %>%
    summarise(
      detected = n() > 0,
      # undetected = 24 - detected,
      snvs = n()
    )
}, simplify = FALSE)
```

```{r}
sapply(deGenes, function(x){
  lapply(topASE, function(y){
    dplyr::filter(y, feat %in% x$gene_id)
  }) %>%
    bind_rows() %>%
    group_by(feat, genotype) %>%
    summarise(
      tested = length(sample),
      ASE = sum(ASE)
    )
}, simplify = FALSE)
```

# Enrichment testing

## Fisher's exact test

```{r}
fisherRes <- bind_rows(topASE) %>%
  left_join(chrGenes_df) %>%
  group_by(sample, chr) %>%
  summarise(ase_chr = sum(ASE), noAse_chr = sum(!ASE)) %>%
  group_by(sample) %>%
  mutate(
    ase_notChr = sum(ase_chr) - ase_chr,
    noAse_notChr = sum(noAse_chr) - noAse_chr
  ) %>%
  ungroup() %>%
  mutate(
    fisherTest = pmap(
      ## Setup list input to pmap in order that will construct a correct matrix
      list(ase_chr, ase_notChr, noAse_chr, noAse_notChr),
      ~ fisher.test(matrix(c(..1, ..2, ..3, ..4), ncol = 2))
    ),
    fisher.p = vapply(
      fisherTest,
      function(x){x$p.value},
      numeric(1)
    ),
    fdr.p = p.adjust(fisher.p, method = "fdr"),
    bonf.p = p.adjust(fisher.p, method = "bonf")
  )
```

## Gene Set Enrichment Analysis

### Setup

```{r}
wilsonConf <- function(p.vals, conf.level = 0.95){
  p <- mean(p.vals)
  n <- length(p.vals)
  alpha <- (1 - conf.level)
  alpha2 <- alpha / 2
  z <- qnorm(1 - alpha2)
  z2 <- z * z
  p1 <- p + 0.5 * z2/n
  p2 <- z * sqrt((p * (1 - p) + 0.25 * z2/n)/n)
  p3 <- 1 + z2/n
  lcl <- (p1 - p2)/p3
  ucl <- (p1 + p2)/p3
  res.wilson <- tibble(
    mean.p = p,
    lower = lcl,
    upper = ucl,
    n = n,
    sig = mean.p < alpha
  )
  return(res.wilson)
}
```

```{r}
bonfConf <- function(adj.p.val, n, n.bonf, conf.level = 0.95, limit = c("lower", "upper")){
  alpha <- ((1 - conf.level) / n.bonf)
  alpha2 <- alpha / 2
  z <- qnorm(1 - alpha2)
  z2 <- z * z
  p1 <- adj.p.val + 0.5 * z2/n
  p2 <- z * sqrt((adj.p.val * (1 - adj.p.val) + 0.25 * z2/n)/n)
  p3 <- 1 + z2/n
  lcl <- (p1 - p2)/p3
  ucl <- (p1 + p2)/p3
  if (limit == "lower"){return(lcl)}
  if (limit == "upper"){return(ucl)}
}
```

```{r}
intSizeByN <- lapply(seq(1:1000), function(n){
  p <- 0.05
  alpha <- 1 - p
  alpha2 <- 0.5 * alpha
  z <- qnorm(1 - alpha2)
  z2 <- z * z
  p1 <- p + 0.5 * z2/n
  p2 <- z * sqrt((p * (1 - p) + 0.25 * z2/n)/n)
  p3 <- 1 + z2/n
  lcl <- (p1 - p2)/p3
  ucl <- (p1 + p2)/p3
  res.wilson <- tibble(
    p = p,
    n = n,
    lower = lcl,
    upper = ucl
  )
  return(res.wilson)
})
intSizeByN %>%
  bind_rows() %>%
  mutate(size = upper - lower) %>%
  ggplot(aes(n, size)) +
  geom_line() +
  labs(y = "Wilson interval size") +
  scale_y_continuous(breaks = seq(0, 0.03, by = 0.001)) +
  scale_x_continuous(breaks = seq(0, 1000, by = 100))
```

### Non-directional

```{r}
permRanks <- function(topASE){
  topASE[,c("feat", "fdr")] %>%
    mutate(
      fdr = ifelse(fdr == 0, 0.000001, fdr),
      stat = -log10(fdr)
    ) %>%
    split(f = .$stat) %>%
    lapply(function(x){
      pull(x, feat) %>%
        sample(length(.), replace = FALSE)
    }) %>%
    unlist(use.names = FALSE) %>%
    enframe(name = NULL, value = "geneid") %>%
    mutate(stat = seq(1:nrow(.))) %>%
    with(structure(stat, names = geneid)) %>%
    rev()
}
```

```{r}
perms <- 500
fgseaFile <- paste0(
  here(),
  "/files/fgsea_",
  perms,
  "perms.Rds"
)
```

```{r}
## This code chunk has a lengthy run-time and will not run if the object has already been saved
## Due to its large size the object remains on the HPC system
## Across 6 threads it takes 1-2 hours to complete
if (!file.exists(fgseaFile)) {
  plan(list(tweak(multisession, workers = cores), sequential))
  ranks <- future_sapply(1:perms, function(...){
    future_sapply(topASE, permRanks, future.seed = 17)
  }, future.seed = 17) %>%
    t() %>%
    as_tibble()
  fgsea <- future_lapply(as.list(ranks), function(x){
    future_lapply(x, function(y){
      fgseaMultilevel(chrGenes, y, scoreType = "pos", eps = 0)
    }, future.seed = 17)
  }, future.seed = 17) %>%
    set_names(colnames(ranks))
  saveRDS(fgsea, fgseaFile)
  plan(sequential)
}
```

```{r}
fgsea <- readRDS(fgseaFile)
```

```{r}
pvals <- lapply(fgsea, function(x){
  lapply(seq_along(x), function(y){
    x[[y]][,c("pathway", "padj")] %>%
      dplyr::rename(!!as.character(y) := padj)
  }) %>%
    purrr::reduce(left_join) %>%
    column_to_rownames("pathway") %>%
    t() %>%
    as_tibble()
})
```

```{r}
mean <- lapply(pvals, function(sample){
  chrList <- as.list(sample)
  lapply(seq_along(chrList), function(x){
    chr <- names(chrList)[x]
    wilsonConf(chrList[[x]]) %>%
      mutate(chromosome = as.factor(chr))
  }) %>%
    bind_rows() %>%
    dplyr::select(chromosome, everything()) %>%
    mutate(
      adj.mean.p = p.adjust(mean.p, method = "bonf"),
      adj.lower = bonfConf(adj.mean.p, n, length(adj.mean.p), limit = "lower"),
      adj.upper = bonfConf(adj.mean.p, n, length(adj.mean.p), limit = "upper")
    )
}) 
```

```{r, fig.height=7.5}
lapply(seq_along(mean), function(ind){
  x <- mean[[ind]]
  x %>%
    mutate(
      sample = names(mean)[ind],
      chromosome = fct_relevel(chromosome, str_sort(chromosome, numeric = TRUE))
    )
}) %>%
  bind_rows() %>%
  left_join(metadata[,c("alias", "genotype")], by = c("sample" = "alias")) %>%
  ggplot(aes(sample, -log10(adj.mean.p))) +
  geom_bar(stat = "identity", aes(fill = genotype)) +
  geom_errorbar(aes(
    ymax = -log10(adj.lower),
    ymin = -log10(adj.upper)
  ), width = 0.2) +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  facet_wrap(~ chromosome) +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    legend.position = "bottom"
  ) +
  scale_fill_manual(values = genoCols) +
  labs(y = log10plab)
```
