---
title: "Differential Gene Expression Analysis"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE, 
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Setup 

```{r packages}
library(tidyverse)
library(magrittr)
library(parallel)
library(here)
library(AnnotationHub)
library(purrr)
library(scales)
library(kableExtra)
library(edgeR)
library(ggrepel)
library(fgsea)
library(pheatmap)
```

```{r options}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
cores <- detectCores() - 1
```

```{r}
source(here("R/lbFuncs.R"))
```

## Annotations

```{r}
ah <- AnnotationHub() %>%
  subset(species == "Danio rerio") %>%
  subset(rdataclass == "EnsDb")
ensDb <- ah[["AH83189"]] ## Ens101
```

```{r}
transcripts <- transcripts(ensDb)
txLen <- exonsBy(ensDb, "tx") %>%
  width() %>%
  vapply(sum, integer(1))
mcols(transcripts) <- mcols(transcripts)[
  c("tx_id", "gene_id", "gc_content")
] %>%
  as.data.frame() %>%
  mutate(length = txLen)
```

```{r}
geneGc <- transcripts %>%
  mcols() %>%
  as_tibble() %>%
  group_by(gene_id) %>%
  summarise(
    gc_content = sum(gc_content*length) / sum(length),
    length = ceiling(median(length))
  )
```

```{r}
genes <- genes(ensDb)
mcols(genes) <- mcols(genes)[
  c("gene_id", "gene_name", "gene_biotype", "entrezid")
] %>%
  as.data.frame() %>%
  left_join(geneGc)
```

## Count data

```{r}
# counts <- read_tsv(here("03_align/featureCounts/genes.out")) %>%
counts <- read_tsv(here("/hpcfs/users/a1647910/210408_psen1_fADfAI_snv/03_align/featureCounts/genes.out")) %>%
  set_colnames(basename(colnames(.))) %>%
  set_colnames(str_remove(colnames(.), "Aligned.sortedByCoord.out.bam"))
```

```{r}
metadata <- read_csv(here("files/samples.csv")) %>%
  dplyr::select(-sample) %>%
  dplyr::rename(sample = basename, genotype = Genotype) %>%
  ## We need some sample aliases that follow R naming conventions
  mutate(
    alias = c(
      paste0(rep("fAD", 7), seq(1, 7)),
      paste0(rep("fAI", 8), seq(1, 8)),
      paste0(rep("wt", 9), seq(1, 9))
    )
  )
metadata$genotype <- fct_relevel(
  metadata$genotype,
  c("WT", "EOfAD-like/+", "fAI-like/+")
)
```

```{r}
minCPM <- 1.5
minSamples <- 7
dgeList <- counts %>%
  as.data.frame() %>%
  column_to_rownames("Geneid") %>%
  .[rowSums(cpm(.) >= minCPM) >= minSamples,] %>%
  DGEList(
    samples = tibble(sample = colnames(.)) %>%
      left_join(metadata),
    genes = genes[rownames(.),] %>%
      as_tibble() %>%
      dplyr::select(chromosome = seqnames, everything())
  ) %>%
  calcNormFactors()
```

Gene-level count data was loaded from `featureCounts` and imported into `R` as a `DGEList` object.
Genes were removed from the counts matrix if they had at least `r minCPM` CPM in more than `r ncol(dgeList) - minSamples` samples.
`r nrow(dgeList)` genes remained for further analysis with library sizes ranging between `r comma(min(dgeList$samples$lib.size))` and `r comma(max(dgeList$samples$lib.size))`.

```{r, fig.cap="*Expression density plots for all samples after filtering*"}
cpm(dgeList, log = TRUE) %>%
  as.data.frame() %>%
  pivot_longer(
    cols = everything(),
    names_to = "sample",
    values_to = "logCPM"
  ) %>%
  left_join(metadata) %>%
  ggplot(aes(logCPM, colour = genotype, group = sample)) +
  geom_density() +
  labs(
    x = "logCPM",
    y = "Density",
    colour = "Genotype"
  )
```

# DE Analysis

## PCA

```{r}
pca <- dgeList %>%
  cpm(log = TRUE) %>%
  t() %>%
  prcomp() 
pcaVars <- percent_format(0.1)(summary(pca)$importance["Proportion of Variance",])
```

```{r, fig.cap="*Principal Component Analysis of gene-level counts*"}
pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(metadata) %>%
  as_tibble() %>%
  ggplot(aes(PC1, PC2, colour = genotype, fill = genotype, label = alias)) +
  geom_point(size = 2) +
  geom_text_repel(show.legend = FALSE) +
  stat_ellipse(geom = "polygon", alpha = 0.05, show.legend = FALSE) +
  guides(fill = FALSE) +
  labs(
    x = paste0("PC1 (", pcaVars[["PC1"]], ")"),
    y = paste0("PC2 (", pcaVars[["PC2"]], ")"),
    colour = "Genotype"
  )
```

A Principal Component Analysis (PCA) was performed using the sample logCPM values.
There is no clustering of genotypes, which may be due to the subtlty of the mutations and early time point (6 months).
Additionally, PC1 only accounts for `r pcaVars[["PC1"]]` of the variance.

## Design

The design matrix was defined containing an intercept representing baseline expression in wildtype samples. 
The remaining two columns represent the presence of either the EofAD or fAI mutant alleles.

```{r}
design <- model.matrix(~genotype, data = dgeList$samples) %>%
  set_colnames(str_remove(colnames(.), "genotype"))
```

```{r, fig.height=6, fig.width=6, fig.cap = "*Visualisation of the design matrix*"}
pheatmap(
  design, 
  cluster_cols = FALSE, 
  cluster_rows = FALSE, 
  color = c("white", "grey50"),
  annotation_row = dgeList$samples["genotype"],
  legend = FALSE
)
```

## Model fitting

```{r}
alpha <- 0.05
minLfc <- 1
fit <- estimateDisp(dgeList, design) %>%
  glmFit()
topTables <- colnames(design)[2:3] %>%
  sapply(function(x){
    glmLRT(fit, coef = x) %>%
      topTags(n = Inf) %>%
      .[["table"]] %>%
      as_tibble() %>%
      arrange(PValue) %>%
      dplyr::select(
        gene_id, gene_name, logFC, logCPM, PValue, FDR, everything()  
      ) %>%
      mutate(
        coef = x,
        bonfP = p.adjust(PValue, "bonf"),
        DE = case_when(
          bonfP < alpha ~ TRUE,
          FDR < alpha & abs(logFC) > minLfc ~ TRUE
        ),
        DE = ifelse(is.na(DE), FALSE, DE)
      )
  }, simplify = FALSE)
```

## Bias checks

```{r}
topTables %>%
  bind_rows() %>%
  mutate(stat = -sign(logFC)*log10(PValue)) %>%
  ggplot(aes(gc_content, stat)) +
  geom_point(aes(colour = DE), alpha = 0.4) +
  geom_smooth(se = FALSE) +
  facet_wrap(~coef)  +
  labs(
    x = "GC content (%)",
    y = "Ranking Statistic"
  ) +
  coord_cartesian(ylim = c(-10, 10)) +
  # scale_colour_manual(values = deCols) +
  theme(legend.position = "none")
```

