---
title: "Variants Analysis"
subtitle: "210408_psen1_fADfAI dataset"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    theme: flatly
    toc: true
    toc_float: true
    toc_depth: 4
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE, 
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Setup 

```{r packages}
suppressPackageStartupMessages({
  ## Common
  library(tidyverse)
  library(magrittr)
  library(future.apply)
  library(here)
  library(AnnotationHub)
  library(purrr)
  library(scales)
  library(kableExtra)
  library(tictoc)
  library(ggrepel)
  library(RColorBrewer)
  library(ggpubr)
  library(pander)
  library(rmarkdown)
  ## Project specific
  library(VariantAnnotation)
  library(Gviz)
  library(zoo)
})
```

```{r options}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
cores <- availableCores() - 1
```

```{r}
# source("~/bioinformatics/bioToolkit/lbFuncs.R")
```

```{r}
## Choose either local or remote project directory
# projDir <- here()
projDir <- "/hpcfs/users/a1647910/210408_psen1_fADfAI"
```

```{r}
options(ucscChromosomeNames = FALSE)
```

## EnsDb

```{r}
ah <- AnnotationHub() %>%
  subset(species == "Danio rerio") %>%
  subset(rdataclass == "EnsDb")
ensDb <- ah[["AH83189"]] ## Ens101
genes <- genes(ensDb, filter = SeqNameFilter(drChrs))
mcols(genes) <- mcols(genes)[
  c("gene_id", "gene_name", "gene_biotype", "entrezid")
]
exons <- exonsBy(ensDb, by = "gene", filter = SeqNameFilter(drChrs))
```

An `EnsDb` object for Ensembl release 101 was accessed and gene and exon annotation information was extracted.

## Metadata

```{r}
metadata <- read_csv(file.path(projDir, "files/samples.csv")) %>%
  dplyr::select(-sample) %>%
  mutate(
    Genotype = factor(Genotype, levels = unique(Genotype)),
    basename = factor(basename, levels = basename),
    alias = c(
      paste0(rep("fAD", 7), seq(1, 7)),
      paste0(rep("fAI", 8), seq(1, 8)),
      paste0(rep("wt", 9), seq(1, 9))
    )
  ) %>%
  dplyr::select(sample = basename, fish_id, batch_killed, sex = Sex, RINe,
                genotype = Genotype, genotype_2 = Genotype_2, alias)
metadata %>%
  dplyr::select(
    Sample = sample, Genotype = genotype, Alias = alias, Gender = sex,
    `Fish ID` = fish_id, `Batch Killed` = batch_killed
  ) %>%
  kable(
    align = "l",
    caption = "Sample metadata"
  ) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed", "responsive")
  )
genoCols <- metadata$genotype %>%
  levels() %>%
  length() %>%
  brewer.pal("Set1") %>%
  setNames(levels(metadata$genotype))
compCols <- genoCols[1:2]
```

```{r}
samples_byGeno <- metadata %>%
  split(f = .$genotype) %>%
  sapply(function(x){
    pull(x, sample) %>%
      droplevels()
  }, simplify = FALSE)
alz_samples <- metadata[metadata$genotype == "EOfAD-like/+",]$sample %>%
  droplevels()
acne_samples <- metadata[metadata$genotype == "fAI-like/+",]$sample %>%
  droplevels()
wt_samples <- metadata[metadata$genotype == "WT",]$sample %>%
  droplevels()
```

```{r}
drChrs <- seq(1:25)
```

# DE data

```{r deData}
topDE <- readRDS(file.path(projDir, "files/topTables_cqn.Rds")) %>%
  set_names(c("fAD", "fAI"))
```

```{r deGenes}
deGenes <- sapply(topDE, dplyr::filter, DE, simplify = FALSE)
de_fAD <- deGenes$fAD$gene_id
de_fAI <- deGenes$fAI$gene_id
```

# Variation

```{r}
vcf_file <- file.path(
  "/hpcfs/users/a1647910/210408_psen1_fADfAI",
  "analysis-variants/10_variants/6_select/all_samples.vcf.gz"
)
svp <- ScanVcfParam(info = "", geno = c("GT", "GQ"))
vcf <- suppressWarnings({
  readVcf(vcf_file, param = svp)
})
```

```{r}
gr <- rowRanges(vcf)
mcols(gr) <- NULL
```

```{r}
unphase_GTs <- function(x){
  str_replace(x, "\\|", "\\/")
}
gt <- geno(vcf)$GT %>%
  as.data.frame() %>%
  mutate(across(everything(), unphase_GTs))
```

## Within groups
<!--
```{r}
gt_similarity <- lapply(samples_byGeno, function(samples){
gt_summary <- apply(gt[,samples], 1, function(x){
## Sort the vector so that the "./." (no-call) genotype is always last
## which.max() will therefore always return a called genotype if tied with a no-call
tab <- table(x) %>%
.[rev(sort(names(.)))]
gt_max <- names(which.max(tab))
n_max <- sum(x == gt_max)
n_called <- sum(x != "./.")
n_nocall <- sum(x == "./.")
tibble(
gt_max = gt_max,
n_max = n_max,
n_called = n_called,
n_nocall = n_nocall
)
}) %>%
bind_rows()
mcols(gr) <- gt_summary
gr %>%
plyranges::filter(gt_max != "./.") %>%  # Because of the above sort, this won't remove rows where n_nocall == most called genotype
plyranges::mutate(
percent = n_max / n_called,
cumsum = cumsum(percent),
lag = dplyr::lag(cumsum, n = 50, default = NA),
lead = dplyr::lead(cumsum, n = 50, default = NA),
window_sum = lead - lag,
window_ave = window_sum / 101
)
})
```

```{r}
lapply(seq_along(gt_similarity), function(ind){
overlaps <- findOverlaps(gt_similarity[[ind]], gr)
colname <- names(gt_similarity[ind])
mcols(gr)[,colname] <<- NA
mcols(gr)[subjectHits(overlaps), colname] <<-
mcols(gt_similarity[[ind]])[queryHits(overlaps), "window_ave"]
})
```

```{r}
axis_track <- GenomeAxisTrack()
```

```{r}
aTrack_alz <- genes[de_fAD] %>%
plyranges::filter(seqnames == "17") %>%
GeneRegionTrack(
name = "Alz DE",
transcriptAnnotation = "symbol",
fontcolor.group = 1,
cex.group = 0.5,
rotation.group = c(35, 35, 35, 0, 0)
)
dTrack_alz <- gr[,c("EOfAD-like/+", "WT")] %>%
plyranges::filter(seqnames == "17") %>%
DataTrack(
type = "l",
name = "Alzheimer's comparison",
window = -1,
windowSize = 1,
groups = c("EOfAD-like/+","WT"),
col = genoCols[c("EOfAD-like/+", "WT")],
baseline = c(0.5, 1)
)
aTrack_acne <- genes[de_fAI] %>%
plyranges::filter(seqnames == "17") %>%
GeneRegionTrack(
name = "Acne DE",
transcriptAnnotation = "symbol",
fontcolor.group = 1,
cex.group = 0.5
)
dTrack_acne <- gr[,c("fAI-like/+", "WT")] %>%
plyranges::filter(seqnames == "17") %>%
DataTrack(
type = "l",
name = "Acne Inversa Comparison",
window = -1,
windowSize = 1,
groups = c("fAI-like/+","WT"),
col = genoCols[c("fAI-like/+", "WT")]
)
plotTracks(list(axis_track, aTrack_alz, dTrack_alz, aTrack_acne, dTrack_acne))
```

```{r}
axis_track <- GenomeAxisTrack()
aTrack_alz <- genes[de_fAD] %>%
plyranges::filter(seqnames == "6") %>%
GeneRegionTrack(
name = "Alz DE",
transcriptAnnotation = "symbol",
fontcolor.group = 1,
cex.group = 0.5,
rotation.group = c(35, 35, 35, 0, 0)
)
dTrack_alz <- gr[,c("EOfAD-like/+", "WT")] %>%
plyranges::filter(seqnames == "6") %>%
DataTrack(
type = "l",
name = "Alzheimer's comparison",
window = -1,
windowSize = 1,
groups = c("EOfAD-like/+","WT"),
col = genoCols[c("EOfAD-like/+", "WT")]
)
plotTracks(list(axis_track, aTrack_alz, dTrack_alz))
```
-->
## Between groups

```{r}
allele_file <- file.path(here(), "files", "allele_stats.Rds")
```

```{r}
## Takes 15-20 mins to run so .Rds saved for convenience
if (!file.exists(allele_file)) {
  allele_stats <- lapply(samples_byGeno, function(samples){
    apply(gt[,samples], 1, function(genos){
      alleles <- genos %>%
        as.character() %>%
        str_split("/") %>%
        unlist()
      n_called <- sum(alleles != ".")
      n_nocall <- sum(alleles == ".")
      n_0 <- sum(alleles == "0")
      n_1 <- sum(alleles == "1")
      n_2 <- sum(alleles == "2")
      n_3 <- sum(alleles == "3")
      tibble(
        n_called = n_called,
        n_nocall = n_nocall,
        n_0 = n_0,
        n_1 = n_1,
        n_2 = n_2,
        n_3 = n_3,
      )
    }) %>%
      bind_rows()
  })
  saveRDS(object = allele_stats, file = allele_file)
} else {
  allele_stats <- readRDS(allele_file)
}
```

```{r}
allele_props <- lapply(allele_stats, function(x){
  x %>%
    mutate(
      prop_ref = ifelse(n_called > n_nocall, n_0 / n_called, NA),
      prop_called = n_called / (n_called + n_nocall)
    )
})
```

### EOfAD

```{r}
window <- 101
gr_alz <- gr
gr_alz$comp <- abs(allele_props$`EOfAD-like/+`$prop_ref - allele_props$WT$prop_ref)
gr_alz$qual_ave <- (allele_props$`EOfAD-like/+`$prop_called + allele_props$WT$prop_called) / 2
gr_alz <- plyranges::filter(gr_alz, !is.na(comp)) %>%
  plyranges::mutate(
    comp_window = rollmean(comp, k = window, fill = NA),
    qual_window = rollmean(qual_ave, k = window, fill = NA),
  ) %>%
  plyranges::select(-c(comp, qual_ave))
```

```{r}
axis_track <- GenomeAxisTrack()
geneTrack_alz <- genes[de_fAD] %>%
  plyranges::filter(seqnames == "17") %>%
  plyranges::mutate(symbol = gene_name) %>%
  GeneRegionTrack(
    name = "DE genes",
    transcriptAnnotation = "symbol",
    showTranscriptId = TRUE,
    fontcolor.group = 1,
    cex.group = 0.5,
    rotation.group = c(35, 35, 35, 0, 0)
  )
variationTrack_alz <- gr_alz[,"comp_window"] %>%
  plyranges::filter(seqnames == "17") %>%
  DataTrack(
    type = "l",
    name = "Allelic variation Alz vs WT",
    window = -1,
    windowSize = 1,
  )
qualTrack_alz <- gr_alz[,"qual_window"] %>%
  plyranges::filter(seqnames == "17") %>%
  DataTrack(
    type = "l",
    name = "Quality",
    window = -1,
    windowSize = 1,
    color = "black"
  )
plotTracks(list(axis_track, geneTrack_alz, variationTrack_alz, qualTrack_alz))
```

### fAI

```{r}
window <- 101
gr_acne <- gr
gr_acne$comp <- abs(allele_props$`fAI-like/+`$prop_ref - allele_props$WT$prop_ref)
gr_acne$qual_ave <- (allele_props$`fAI-like/+`$prop_called + allele_props$WT$prop_called) / 2
gr_acne <- plyranges::filter(gr_acne, !is.na(comp)) %>%
  plyranges::mutate(
    comp_window = rollmean(comp, k = window, fill = NA),
    qual_window = rollmean(qual_ave, k = window, fill = NA),
  ) %>%
  plyranges::select(-c(comp, qual_ave))
```

```{r}
axis_track <- GenomeAxisTrack()
geneTrack_acne <- genes[de_fAI] %>%
  plyranges::filter(seqnames == "17") %>%
  plyranges::mutate(symbol = gene_name) %>%
  GeneRegionTrack(
    name = "DE genes",
    transcriptAnnotation = "symbol",
    showTranscriptId = TRUE,
    fontcolor.group = 1,
    cex.group = 0.5,
    rotation.group = c(35, 35, 35, 0, 0)
  )
variationTrack_acne <- gr_acne[,"comp_window"] %>%
  plyranges::filter(seqnames == "17") %>%
  DataTrack(
    type = "l",
    name = "Allelic variation Acne vs WT",
    window = -1,
    windowSize = 1,
  )
qualTrack_acne <- gr_acne[,"qual_window"] %>%
  plyranges::filter(seqnames == "17") %>%
  DataTrack(
    type = "l",
    name = "Quality",
    window = -1,
    windowSize = 1,
    color = "black"
  )
plotTracks(list(axis_track, geneTrack_acne, variationTrack_acne, qualTrack_acne))
```

Circos plot:
  - cube root the dissimilarity metric
  - plot absolute logFC
  
Look at regulatory TFT datasets

Try quasi likelihood for DE testing

The window is called an elastic window
  - Give a summary of window sizes